// @pinecone-database/pinecone/dist/inference/rerank.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rerank = void 0;
const inference_1 = require("../pinecone-generated-ts-fetch/inference");
const errors_1 = require("../errors");
const rerank = (infApi) => {
    return async (options) => {
        if (!options.documents || options.documents.length == 0) {
            throw new errors_1.PineconeArgumentError('You must pass at least one document to rerank');
        }
        if (!options.query || options.query.length == 0) {
            throw new errors_1.PineconeArgumentError('You must pass a query to rerank');
        }
        if (!options.model || options.model.length == 0) {
            throw new errors_1.PineconeArgumentError('You must pass the name of a supported reranking model in order to rerank' +
                ' documents. See https://docs.pinecone.io/models for supported models.');
        }
        const topN = options.topN ?? options.documents.length;
        const returnDocuments = options.returnDocuments ?? true;
        const parameters = options.parameters ?? {};
        const rankFields = options.rankFields ?? ['text'];
        // Validate and standardize documents to ensure they are in object format
        const newDocuments = options.documents.map((doc) => typeof doc === 'string' ? { text: doc } : doc);
        if (!options.rankFields) {
            if (!newDocuments.every((doc) => typeof doc === 'object' && doc.text)) {
                throw new errors_1.PineconeArgumentError('Documents must be a list of strings or objects containing the "text" field');
            }
        }
        return await infApi.rerank({
            rerankRequest: {
                model: options.model,
                query: options.query,
                documents: newDocuments,
                topN: topN,
                returnDocuments: returnDocuments,
                rankFields,
                parameters: parameters,
            },
            xPineconeApiVersion: inference_1.X_PINECONE_API_VERSION,
        });
    };
};
exports.rerank = rerank;
