// weaviate-client/dist/node/cjs/grpc/batcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const nice_grpc_1 = require("nice-grpc");
const batch_js_1 = require("../proto/v1/batch.js");
const errors_js_1 = require("../errors.js");
const batch_delete_js_1 = require("../proto/v1/batch_delete.js");
const base_js_1 = __importDefault(require("./base.js"));
const retry_js_1 = require("./retry.js");
class Batcher extends base_js_1.default {
    static use(connection, collection, metadata, timeout, consistencyLevel, tenant) {
        return new Batcher(connection, collection, metadata, timeout, consistencyLevel, tenant);
    }
    withStream = (reqs) => this.callStream(reqs);
    withDelete = (args) => this.callDelete(batch_delete_js_1.BatchDeleteRequest.fromPartial(args));
    withObjects = (args) => this.callObjects(batch_js_1.BatchObjectsRequest.fromPartial(args));
    async *callStream(reqs) {
        const consistencyLevel = this.consistencyLevel;
        async function* generate() {
            yield batch_js_1.BatchStreamRequest.create({ start: { consistencyLevel } });
            for await (const req of reqs) {
                yield req;
            }
        }
        try {
            for await (const res of this.connection.batchStream(generate(), { metadata: this.metadata })) {
                yield res;
            }
        }
        catch (err) {
            if (err instanceof nice_grpc_1.ClientError) {
                throw new errors_js_1.WeaviateBatchStreamError(err.message);
            }
            throw err;
        }
    }
    callDelete(message) {
        return this.sendWithTimeout((signal) => this.connection.batchDelete({
            ...message,
            collection: this.collection,
            consistencyLevel: this.consistencyLevel,
            tenant: this.tenant,
        }, {
            metadata: this.metadata,
            signal,
        }), (err) => new errors_js_1.WeaviateDeleteManyError(err.message));
    }
    callObjects(message) {
        return this.sendWithTimeout((signal) => this.connection.batchObjects({
            ...message,
            consistencyLevel: this.consistencyLevel,
        }, {
            metadata: this.metadata,
            signal,
            ...retry_js_1.retryOptions,
        }), (err) => new errors_js_1.WeaviateBatchError(err.message));
    }
}
exports.default = Batcher;
