// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/BYOC.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfBYOC = instanceOfBYOC;
exports.BYOCFromJSON = BYOCFromJSON;
exports.BYOCFromJSONTyped = BYOCFromJSONTyped;
exports.BYOCToJSON = BYOCToJSON;
const ByocSpecResponse_1 = require("./ByocSpecResponse");
/**
 * Check if a given object implements the BYOC interface.
 */
function instanceOfBYOC(value) {
    let isInstance = true;
    isInstance = isInstance && "byoc" in value;
    return isInstance;
}
function BYOCFromJSON(json) {
    return BYOCFromJSONTyped(json, false);
}
function BYOCFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'byoc': (0, ByocSpecResponse_1.ByocSpecResponseFromJSON)(json['byoc']),
    };
}
function BYOCToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'byoc': (0, ByocSpecResponse_1.ByocSpecResponseToJSON)(value.byoc),
    };
}
