// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/BYOC1.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfBYOC1 = instanceOfBYOC1;
exports.BYOC1FromJSON = BYOC1FromJSON;
exports.BYOC1FromJSONTyped = BYOC1FromJSONTyped;
exports.BYOC1ToJSON = BYOC1ToJSON;
const BYOC1Byoc_1 = require("./BYOC1Byoc");
/**
 * Check if a given object implements the BYOC1 interface.
 */
function instanceOfBYOC1(value) {
    let isInstance = true;
    isInstance = isInstance && "byoc" in value;
    return isInstance;
}
function BYOC1FromJSON(json) {
    return BYOC1FromJSONTyped(json, false);
}
function BYOC1FromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'byoc': (0, BYOC1Byoc_1.BYOC1ByocFromJSON)(json['byoc']),
    };
}
function BYOC1ToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'byoc': (0, BYOC1Byoc_1.BYOC1ByocToJSON)(value.byoc),
    };
}
