// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/BYOC1Byoc.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfBYOC1Byoc = instanceOfBYOC1Byoc;
exports.BYOC1ByocFromJSON = BYOC1ByocFromJSON;
exports.BYOC1ByocFromJSONTyped = BYOC1ByocFromJSONTyped;
exports.BYOC1ByocToJSON = BYOC1ByocToJSON;
const runtime_1 = require("../runtime");
const ReadCapacity_1 = require("./ReadCapacity");
/**
 * Check if a given object implements the BYOC1Byoc interface.
 */
function instanceOfBYOC1Byoc(value) {
    let isInstance = true;
    return isInstance;
}
function BYOC1ByocFromJSON(json) {
    return BYOC1ByocFromJSONTyped(json, false);
}
function BYOC1ByocFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'readCapacity': !(0, runtime_1.exists)(json, 'read_capacity') ? undefined : (0, ReadCapacity_1.ReadCapacityFromJSON)(json['read_capacity']),
    };
}
function BYOC1ByocToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'read_capacity': (0, ReadCapacity_1.ReadCapacityToJSON)(value.readCapacity),
    };
}
