// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ByocSpecResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfByocSpecResponse = instanceOfByocSpecResponse;
exports.ByocSpecResponseFromJSON = ByocSpecResponseFromJSON;
exports.ByocSpecResponseFromJSONTyped = ByocSpecResponseFromJSONTyped;
exports.ByocSpecResponseToJSON = ByocSpecResponseToJSON;
const runtime_1 = require("../runtime");
const MetadataSchema_1 = require("./MetadataSchema");
const ReadCapacityResponse_1 = require("./ReadCapacityResponse");
/**
 * Check if a given object implements the ByocSpecResponse interface.
 */
function instanceOfByocSpecResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "environment" in value;
    isInstance = isInstance && "readCapacity" in value;
    return isInstance;
}
function ByocSpecResponseFromJSON(json) {
    return ByocSpecResponseFromJSONTyped(json, false);
}
function ByocSpecResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'environment': json['environment'],
        'readCapacity': (0, ReadCapacityResponse_1.ReadCapacityResponseFromJSON)(json['read_capacity']),
        'schema': !(0, runtime_1.exists)(json, 'schema') ? undefined : (0, MetadataSchema_1.MetadataSchemaFromJSON)(json['schema']),
    };
}
function ByocSpecResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'environment': value.environment,
        'read_capacity': (0, ReadCapacityResponse_1.ReadCapacityResponseToJSON)(value.readCapacity),
        'schema': (0, MetadataSchema_1.MetadataSchemaToJSON)(value.schema),
    };
}
