// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/CollectionList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCollectionList = instanceOfCollectionList;
exports.CollectionListFromJSON = CollectionListFromJSON;
exports.CollectionListFromJSONTyped = CollectionListFromJSONTyped;
exports.CollectionListToJSON = CollectionListToJSON;
const runtime_1 = require("../runtime");
const CollectionModel_1 = require("./CollectionModel");
/**
 * Check if a given object implements the CollectionList interface.
 */
function instanceOfCollectionList(value) {
    let isInstance = true;
    return isInstance;
}
function CollectionListFromJSON(json) {
    return CollectionListFromJSONTyped(json, false);
}
function CollectionListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'collections': !(0, runtime_1.exists)(json, 'collections') ? undefined : (json['collections'].map(CollectionModel_1.CollectionModelFromJSON)),
    };
}
function CollectionListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'collections': value.collections === undefined ? undefined : (value.collections.map(CollectionModel_1.CollectionModelToJSON)),
    };
}
