// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ConfigureIndexRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfConfigureIndexRequest = instanceOfConfigureIndexRequest;
exports.ConfigureIndexRequestFromJSON = ConfigureIndexRequestFromJSON;
exports.ConfigureIndexRequestFromJSONTyped = ConfigureIndexRequestFromJSONTyped;
exports.ConfigureIndexRequestToJSON = ConfigureIndexRequestToJSON;
const runtime_1 = require("../runtime");
const ConfigureIndexRequestEmbed_1 = require("./ConfigureIndexRequestEmbed");
const ConfigureIndexRequestSpec_1 = require("./ConfigureIndexRequestSpec");
/**
 * Check if a given object implements the ConfigureIndexRequest interface.
 */
function instanceOfConfigureIndexRequest(value) {
    let isInstance = true;
    return isInstance;
}
function ConfigureIndexRequestFromJSON(json) {
    return ConfigureIndexRequestFromJSONTyped(json, false);
}
function ConfigureIndexRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'spec': !(0, runtime_1.exists)(json, 'spec') ? undefined : (0, ConfigureIndexRequestSpec_1.ConfigureIndexRequestSpecFromJSON)(json['spec']),
        'deletionProtection': !(0, runtime_1.exists)(json, 'deletion_protection') ? undefined : json['deletion_protection'],
        'tags': !(0, runtime_1.exists)(json, 'tags') ? undefined : json['tags'],
        'embed': !(0, runtime_1.exists)(json, 'embed') ? undefined : (0, ConfigureIndexRequestEmbed_1.ConfigureIndexRequestEmbedFromJSON)(json['embed']),
    };
}
function ConfigureIndexRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'spec': (0, ConfigureIndexRequestSpec_1.ConfigureIndexRequestSpecToJSON)(value.spec),
        'deletion_protection': value.deletionProtection,
        'tags': value.tags,
        'embed': (0, ConfigureIndexRequestEmbed_1.ConfigureIndexRequestEmbedToJSON)(value.embed),
    };
}
