// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ConfigureIndexRequestEmbed.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfConfigureIndexRequestEmbed = instanceOfConfigureIndexRequestEmbed;
exports.ConfigureIndexRequestEmbedFromJSON = ConfigureIndexRequestEmbedFromJSON;
exports.ConfigureIndexRequestEmbedFromJSONTyped = ConfigureIndexRequestEmbedFromJSONTyped;
exports.ConfigureIndexRequestEmbedToJSON = ConfigureIndexRequestEmbedToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ConfigureIndexRequestEmbed interface.
 */
function instanceOfConfigureIndexRequestEmbed(value) {
    let isInstance = true;
    return isInstance;
}
function ConfigureIndexRequestEmbedFromJSON(json) {
    return ConfigureIndexRequestEmbedFromJSONTyped(json, false);
}
function ConfigureIndexRequestEmbedFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'model': !(0, runtime_1.exists)(json, 'model') ? undefined : json['model'],
        'fieldMap': !(0, runtime_1.exists)(json, 'field_map') ? undefined : json['field_map'],
        'readParameters': !(0, runtime_1.exists)(json, 'read_parameters') ? undefined : json['read_parameters'],
        'writeParameters': !(0, runtime_1.exists)(json, 'write_parameters') ? undefined : json['write_parameters'],
    };
}
function ConfigureIndexRequestEmbedToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'model': value.model,
        'field_map': value.fieldMap,
        'read_parameters': value.readParameters,
        'write_parameters': value.writeParameters,
    };
}
