// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/CreateIndexFromBackupRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateIndexFromBackupRequest = instanceOfCreateIndexFromBackupRequest;
exports.CreateIndexFromBackupRequestFromJSON = CreateIndexFromBackupRequestFromJSON;
exports.CreateIndexFromBackupRequestFromJSONTyped = CreateIndexFromBackupRequestFromJSONTyped;
exports.CreateIndexFromBackupRequestToJSON = CreateIndexFromBackupRequestToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the CreateIndexFromBackupRequest interface.
 */
function instanceOfCreateIndexFromBackupRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "name" in value;
    return isInstance;
}
function CreateIndexFromBackupRequestFromJSON(json) {
    return CreateIndexFromBackupRequestFromJSONTyped(json, false);
}
function CreateIndexFromBackupRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': json['name'],
        'tags': !(0, runtime_1.exists)(json, 'tags') ? undefined : json['tags'],
        'deletionProtection': !(0, runtime_1.exists)(json, 'deletion_protection') ? undefined : json['deletion_protection'],
    };
}
function CreateIndexFromBackupRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'tags': value.tags,
        'deletion_protection': value.deletionProtection,
    };
}
