// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/CreateIndexForModelRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateIndexForModelRequest = instanceOfCreateIndexForModelRequest;
exports.CreateIndexForModelRequestFromJSON = CreateIndexForModelRequestFromJSON;
exports.CreateIndexForModelRequestFromJSONTyped = CreateIndexForModelRequestFromJSONTyped;
exports.CreateIndexForModelRequestToJSON = CreateIndexForModelRequestToJSON;
const runtime_1 = require("../runtime");
const CreateIndexForModelRequestEmbed_1 = require("./CreateIndexForModelRequestEmbed");
const MetadataSchema_1 = require("./MetadataSchema");
const ReadCapacity_1 = require("./ReadCapacity");
/**
 * Check if a given object implements the CreateIndexForModelRequest interface.
 */
function instanceOfCreateIndexForModelRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "name" in value;
    isInstance = isInstance && "cloud" in value;
    isInstance = isInstance && "region" in value;
    isInstance = isInstance && "embed" in value;
    return isInstance;
}
function CreateIndexForModelRequestFromJSON(json) {
    return CreateIndexForModelRequestFromJSONTyped(json, false);
}
function CreateIndexForModelRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': json['name'],
        'cloud': json['cloud'],
        'region': json['region'],
        'deletionProtection': !(0, runtime_1.exists)(json, 'deletion_protection') ? undefined : json['deletion_protection'],
        'tags': !(0, runtime_1.exists)(json, 'tags') ? undefined : json['tags'],
        'schema': !(0, runtime_1.exists)(json, 'schema') ? undefined : (0, MetadataSchema_1.MetadataSchemaFromJSON)(json['schema']),
        'readCapacity': !(0, runtime_1.exists)(json, 'read_capacity') ? undefined : (0, ReadCapacity_1.ReadCapacityFromJSON)(json['read_capacity']),
        'embed': (0, CreateIndexForModelRequestEmbed_1.CreateIndexForModelRequestEmbedFromJSON)(json['embed']),
    };
}
function CreateIndexForModelRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'cloud': value.cloud,
        'region': value.region,
        'deletion_protection': value.deletionProtection,
        'tags': value.tags,
        'schema': (0, MetadataSchema_1.MetadataSchemaToJSON)(value.schema),
        'read_capacity': (0, ReadCapacity_1.ReadCapacityToJSON)(value.readCapacity),
        'embed': (0, CreateIndexForModelRequestEmbed_1.CreateIndexForModelRequestEmbedToJSON)(value.embed),
    };
}
