// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/CreateIndexRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateIndexRequest = instanceOfCreateIndexRequest;
exports.CreateIndexRequestFromJSON = CreateIndexRequestFromJSON;
exports.CreateIndexRequestFromJSONTyped = CreateIndexRequestFromJSONTyped;
exports.CreateIndexRequestToJSON = CreateIndexRequestToJSON;
const runtime_1 = require("../runtime");
const IndexSpec_1 = require("./IndexSpec");
/**
 * Check if a given object implements the CreateIndexRequest interface.
 */
function instanceOfCreateIndexRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "name" in value;
    isInstance = isInstance && "spec" in value;
    return isInstance;
}
function CreateIndexRequestFromJSON(json) {
    return CreateIndexRequestFromJSONTyped(json, false);
}
function CreateIndexRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': json['name'],
        'dimension': !(0, runtime_1.exists)(json, 'dimension') ? undefined : json['dimension'],
        'metric': !(0, runtime_1.exists)(json, 'metric') ? undefined : json['metric'],
        'deletionProtection': !(0, runtime_1.exists)(json, 'deletion_protection') ? undefined : json['deletion_protection'],
        'tags': !(0, runtime_1.exists)(json, 'tags') ? undefined : json['tags'],
        'spec': (0, IndexSpec_1.IndexSpecFromJSON)(json['spec']),
        'vectorType': !(0, runtime_1.exists)(json, 'vector_type') ? undefined : json['vector_type'],
    };
}
function CreateIndexRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'dimension': value.dimension,
        'metric': value.metric,
        'deletion_protection': value.deletionProtection,
        'tags': value.tags,
        'spec': (0, IndexSpec_1.IndexSpecToJSON)(value.spec),
        'vector_type': value.vectorType,
    };
}
