// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/CreateIndexFromBackupResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateIndexFromBackupResponse = instanceOfCreateIndexFromBackupResponse;
exports.CreateIndexFromBackupResponseFromJSON = CreateIndexFromBackupResponseFromJSON;
exports.CreateIndexFromBackupResponseFromJSONTyped = CreateIndexFromBackupResponseFromJSONTyped;
exports.CreateIndexFromBackupResponseToJSON = CreateIndexFromBackupResponseToJSON;
/**
 * Check if a given object implements the CreateIndexFromBackupResponse interface.
 */
function instanceOfCreateIndexFromBackupResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "restoreJobId" in value;
    isInstance = isInstance && "indexId" in value;
    return isInstance;
}
function CreateIndexFromBackupResponseFromJSON(json) {
    return CreateIndexFromBackupResponseFromJSONTyped(json, false);
}
function CreateIndexFromBackupResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'restoreJobId': json['restore_job_id'],
        'indexId': json['index_id'],
    };
}
function CreateIndexFromBackupResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'restore_job_id': value.restoreJobId,
        'index_id': value.indexId,
    };
}
