// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/IndexModelStatus.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfIndexModelStatus = instanceOfIndexModelStatus;
exports.IndexModelStatusFromJSON = IndexModelStatusFromJSON;
exports.IndexModelStatusFromJSONTyped = IndexModelStatusFromJSONTyped;
exports.IndexModelStatusToJSON = IndexModelStatusToJSON;
/**
 * Check if a given object implements the IndexModelStatus interface.
 */
function instanceOfIndexModelStatus(value) {
    let isInstance = true;
    isInstance = isInstance && "ready" in value;
    isInstance = isInstance && "state" in value;
    return isInstance;
}
function IndexModelStatusFromJSON(json) {
    return IndexModelStatusFromJSONTyped(json, false);
}
function IndexModelStatusFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'ready': json['ready'],
        'state': json['state'],
    };
}
function IndexModelStatusToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'ready': value.ready,
        'state': value.state,
    };
}
