// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/IndexList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfIndexList = instanceOfIndexList;
exports.IndexListFromJSON = IndexListFromJSON;
exports.IndexListFromJSONTyped = IndexListFromJSONTyped;
exports.IndexListToJSON = IndexListToJSON;
const runtime_1 = require("../runtime");
const IndexModel_1 = require("./IndexModel");
/**
 * Check if a given object implements the IndexList interface.
 */
function instanceOfIndexList(value) {
    let isInstance = true;
    return isInstance;
}
function IndexListFromJSON(json) {
    return IndexListFromJSONTyped(json, false);
}
function IndexListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'indexes': !(0, runtime_1.exists)(json, 'indexes') ? undefined : (json['indexes'].map(IndexModel_1.IndexModelFromJSON)),
    };
}
function IndexListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'indexes': value.indexes === undefined ? undefined : (value.indexes.map(IndexModel_1.IndexModelToJSON)),
    };
}
