// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/MetadataSchema.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMetadataSchema = instanceOfMetadataSchema;
exports.MetadataSchemaFromJSON = MetadataSchemaFromJSON;
exports.MetadataSchemaFromJSONTyped = MetadataSchemaFromJSONTyped;
exports.MetadataSchemaToJSON = MetadataSchemaToJSON;
const runtime_1 = require("../runtime");
const MetadataSchemaFieldsValue_1 = require("./MetadataSchemaFieldsValue");
/**
 * Check if a given object implements the MetadataSchema interface.
 */
function instanceOfMetadataSchema(value) {
    let isInstance = true;
    isInstance = isInstance && "fields" in value;
    return isInstance;
}
function MetadataSchemaFromJSON(json) {
    return MetadataSchemaFromJSONTyped(json, false);
}
function MetadataSchemaFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fields': ((0, runtime_1.mapValues)(json['fields'], MetadataSchemaFieldsValue_1.MetadataSchemaFieldsValueFromJSON)),
    };
}
function MetadataSchemaToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fields': ((0, runtime_1.mapValues)(value.fields, MetadataSchemaFieldsValue_1.MetadataSchemaFieldsValueToJSON)),
    };
}
