// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/MetadataSchemaFieldsValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMetadataSchemaFieldsValue = instanceOfMetadataSchemaFieldsValue;
exports.MetadataSchemaFieldsValueFromJSON = MetadataSchemaFieldsValueFromJSON;
exports.MetadataSchemaFieldsValueFromJSONTyped = MetadataSchemaFieldsValueFromJSONTyped;
exports.MetadataSchemaFieldsValueToJSON = MetadataSchemaFieldsValueToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the MetadataSchemaFieldsValue interface.
 */
function instanceOfMetadataSchemaFieldsValue(value) {
    let isInstance = true;
    return isInstance;
}
function MetadataSchemaFieldsValueFromJSON(json) {
    return MetadataSchemaFieldsValueFromJSONTyped(json, false);
}
function MetadataSchemaFieldsValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'filterable': !(0, runtime_1.exists)(json, 'filterable') ? undefined : json['filterable'],
    };
}
function MetadataSchemaFieldsValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'filterable': value.filterable,
    };
}
