// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/PodBased1.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfPodBased1 = instanceOfPodBased1;
exports.PodBased1FromJSON = PodBased1FromJSON;
exports.PodBased1FromJSONTyped = PodBased1FromJSONTyped;
exports.PodBased1ToJSON = PodBased1ToJSON;
const PodBased1Pod_1 = require("./PodBased1Pod");
/**
 * Check if a given object implements the PodBased1 interface.
 */
function instanceOfPodBased1(value) {
    let isInstance = true;
    isInstance = isInstance && "pod" in value;
    return isInstance;
}
function PodBased1FromJSON(json) {
    return PodBased1FromJSONTyped(json, false);
}
function PodBased1FromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'pod': (0, PodBased1Pod_1.PodBased1PodFromJSON)(json['pod']),
    };
}
function PodBased1ToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'pod': (0, PodBased1Pod_1.PodBased1PodToJSON)(value.pod),
    };
}
