// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/PodBased1Pod.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfPodBased1Pod = instanceOfPodBased1Pod;
exports.PodBased1PodFromJSON = PodBased1PodFromJSON;
exports.PodBased1PodFromJSONTyped = PodBased1PodFromJSONTyped;
exports.PodBased1PodToJSON = PodBased1PodToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the PodBased1Pod interface.
 */
function instanceOfPodBased1Pod(value) {
    let isInstance = true;
    return isInstance;
}
function PodBased1PodFromJSON(json) {
    return PodBased1PodFromJSONTyped(json, false);
}
function PodBased1PodFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'replicas': !(0, runtime_1.exists)(json, 'replicas') ? undefined : json['replicas'],
        'podType': !(0, runtime_1.exists)(json, 'pod_type') ? undefined : json['pod_type'],
    };
}
function PodBased1PodToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'replicas': value.replicas,
        'pod_type': value.podType,
    };
}
