// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/PodSpecMetadataConfig.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfPodSpecMetadataConfig = instanceOfPodSpecMetadataConfig;
exports.PodSpecMetadataConfigFromJSON = PodSpecMetadataConfigFromJSON;
exports.PodSpecMetadataConfigFromJSONTyped = PodSpecMetadataConfigFromJSONTyped;
exports.PodSpecMetadataConfigToJSON = PodSpecMetadataConfigToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the PodSpecMetadataConfig interface.
 */
function instanceOfPodSpecMetadataConfig(value) {
    let isInstance = true;
    return isInstance;
}
function PodSpecMetadataConfigFromJSON(json) {
    return PodSpecMetadataConfigFromJSONTyped(json, false);
}
function PodSpecMetadataConfigFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'indexed': !(0, runtime_1.exists)(json, 'indexed') ? undefined : json['indexed'],
    };
}
function PodSpecMetadataConfigToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'indexed': value.indexed,
    };
}
