// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ReadCapacityDedicatedConfig.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReadCapacityDedicatedConfig = instanceOfReadCapacityDedicatedConfig;
exports.ReadCapacityDedicatedConfigFromJSON = ReadCapacityDedicatedConfigFromJSON;
exports.ReadCapacityDedicatedConfigFromJSONTyped = ReadCapacityDedicatedConfigFromJSONTyped;
exports.ReadCapacityDedicatedConfigToJSON = ReadCapacityDedicatedConfigToJSON;
const runtime_1 = require("../runtime");
const ScalingConfigManual_1 = require("./ScalingConfigManual");
/**
 * Check if a given object implements the ReadCapacityDedicatedConfig interface.
 */
function instanceOfReadCapacityDedicatedConfig(value) {
    let isInstance = true;
    return isInstance;
}
function ReadCapacityDedicatedConfigFromJSON(json) {
    return ReadCapacityDedicatedConfigFromJSONTyped(json, false);
}
function ReadCapacityDedicatedConfigFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'nodeType': !(0, runtime_1.exists)(json, 'node_type') ? undefined : json['node_type'],
        'scaling': !(0, runtime_1.exists)(json, 'scaling') ? undefined : json['scaling'],
        'manual': !(0, runtime_1.exists)(json, 'manual') ? undefined : (0, ScalingConfigManual_1.ScalingConfigManualFromJSON)(json['manual']),
    };
}
function ReadCapacityDedicatedConfigToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'node_type': value.nodeType,
        'scaling': value.scaling,
        'manual': (0, ScalingConfigManual_1.ScalingConfigManualToJSON)(value.manual),
    };
}
