// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ReadCapacityOnDemandSpec.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReadCapacityOnDemandSpec = instanceOfReadCapacityOnDemandSpec;
exports.ReadCapacityOnDemandSpecFromJSON = ReadCapacityOnDemandSpecFromJSON;
exports.ReadCapacityOnDemandSpecFromJSONTyped = ReadCapacityOnDemandSpecFromJSONTyped;
exports.ReadCapacityOnDemandSpecToJSON = ReadCapacityOnDemandSpecToJSON;
/**
 * Check if a given object implements the ReadCapacityOnDemandSpec interface.
 */
function instanceOfReadCapacityOnDemandSpec(value) {
    let isInstance = true;
    isInstance = isInstance && "mode" in value;
    return isInstance;
}
function ReadCapacityOnDemandSpecFromJSON(json) {
    return ReadCapacityOnDemandSpecFromJSONTyped(json, false);
}
function ReadCapacityOnDemandSpecFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'mode': json['mode'],
    };
}
function ReadCapacityOnDemandSpecToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'mode': value.mode,
    };
}
