// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ReadCapacityDedicatedSpec.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReadCapacityDedicatedSpec = instanceOfReadCapacityDedicatedSpec;
exports.ReadCapacityDedicatedSpecFromJSON = ReadCapacityDedicatedSpecFromJSON;
exports.ReadCapacityDedicatedSpecFromJSONTyped = ReadCapacityDedicatedSpecFromJSONTyped;
exports.ReadCapacityDedicatedSpecToJSON = ReadCapacityDedicatedSpecToJSON;
const ReadCapacityDedicatedConfig_1 = require("./ReadCapacityDedicatedConfig");
/**
 * Check if a given object implements the ReadCapacityDedicatedSpec interface.
 */
function instanceOfReadCapacityDedicatedSpec(value) {
    let isInstance = true;
    isInstance = isInstance && "mode" in value;
    isInstance = isInstance && "dedicated" in value;
    return isInstance;
}
function ReadCapacityDedicatedSpecFromJSON(json) {
    return ReadCapacityDedicatedSpecFromJSONTyped(json, false);
}
function ReadCapacityDedicatedSpecFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        ...json,
        'mode': json['mode'],
        'dedicated': (0, ReadCapacityDedicatedConfig_1.ReadCapacityDedicatedConfigFromJSON)(json['dedicated']),
    };
}
function ReadCapacityDedicatedSpecToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        ...value,
        'mode': value.mode,
        'dedicated': (0, ReadCapacityDedicatedConfig_1.ReadCapacityDedicatedConfigToJSON)(value.dedicated),
    };
}
