// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ReadCapacityDedicatedSpecResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReadCapacityDedicatedSpecResponse = instanceOfReadCapacityDedicatedSpecResponse;
exports.ReadCapacityDedicatedSpecResponseFromJSON = ReadCapacityDedicatedSpecResponseFromJSON;
exports.ReadCapacityDedicatedSpecResponseFromJSONTyped = ReadCapacityDedicatedSpecResponseFromJSONTyped;
exports.ReadCapacityDedicatedSpecResponseToJSON = ReadCapacityDedicatedSpecResponseToJSON;
const ReadCapacityDedicatedConfig_1 = require("./ReadCapacityDedicatedConfig");
const ReadCapacityStatus_1 = require("./ReadCapacityStatus");
/**
 * Check if a given object implements the ReadCapacityDedicatedSpecResponse interface.
 */
function instanceOfReadCapacityDedicatedSpecResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "mode" in value;
    isInstance = isInstance && "dedicated" in value;
    isInstance = isInstance && "status" in value;
    return isInstance;
}
function ReadCapacityDedicatedSpecResponseFromJSON(json) {
    return ReadCapacityDedicatedSpecResponseFromJSONTyped(json, false);
}
function ReadCapacityDedicatedSpecResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'mode': json['mode'],
        'dedicated': (0, ReadCapacityDedicatedConfig_1.ReadCapacityDedicatedConfigFromJSON)(json['dedicated']),
        'status': (0, ReadCapacityStatus_1.ReadCapacityStatusFromJSON)(json['status']),
    };
}
function ReadCapacityDedicatedSpecResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'mode': value.mode,
        'dedicated': (0, ReadCapacityDedicatedConfig_1.ReadCapacityDedicatedConfigToJSON)(value.dedicated),
        'status': (0, ReadCapacityStatus_1.ReadCapacityStatusToJSON)(value.status),
    };
}
