// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ReadCapacityStatus.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReadCapacityStatus = instanceOfReadCapacityStatus;
exports.ReadCapacityStatusFromJSON = ReadCapacityStatusFromJSON;
exports.ReadCapacityStatusFromJSONTyped = ReadCapacityStatusFromJSONTyped;
exports.ReadCapacityStatusToJSON = ReadCapacityStatusToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ReadCapacityStatus interface.
 */
function instanceOfReadCapacityStatus(value) {
    let isInstance = true;
    isInstance = isInstance && "state" in value;
    return isInstance;
}
function ReadCapacityStatusFromJSON(json) {
    return ReadCapacityStatusFromJSONTyped(json, false);
}
function ReadCapacityStatusFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'state': json['state'],
        'currentReplicas': !(0, runtime_1.exists)(json, 'current_replicas') ? undefined : json['current_replicas'],
        'currentShards': !(0, runtime_1.exists)(json, 'current_shards') ? undefined : json['current_shards'],
        'errorMessage': !(0, runtime_1.exists)(json, 'error_message') ? undefined : json['error_message'],
    };
}
function ReadCapacityStatusToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'state': value.state,
        'current_replicas': value.currentReplicas,
        'current_shards': value.currentShards,
        'error_message': value.errorMessage,
    };
}
