// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ScalingConfigManual.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfScalingConfigManual = instanceOfScalingConfigManual;
exports.ScalingConfigManualFromJSON = ScalingConfigManualFromJSON;
exports.ScalingConfigManualFromJSONTyped = ScalingConfigManualFromJSONTyped;
exports.ScalingConfigManualToJSON = ScalingConfigManualToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ScalingConfigManual interface.
 */
function instanceOfScalingConfigManual(value) {
    let isInstance = true;
    return isInstance;
}
function ScalingConfigManualFromJSON(json) {
    return ScalingConfigManualFromJSONTyped(json, false);
}
function ScalingConfigManualFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'replicas': !(0, runtime_1.exists)(json, 'replicas') ? undefined : json['replicas'],
        'shards': !(0, runtime_1.exists)(json, 'shards') ? undefined : json['shards'],
    };
}
function ScalingConfigManualToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'replicas': value.replicas,
        'shards': value.shards,
    };
}
