// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ServerlessSpecResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfServerlessSpecResponse = instanceOfServerlessSpecResponse;
exports.ServerlessSpecResponseFromJSON = ServerlessSpecResponseFromJSON;
exports.ServerlessSpecResponseFromJSONTyped = ServerlessSpecResponseFromJSONTyped;
exports.ServerlessSpecResponseToJSON = ServerlessSpecResponseToJSON;
const runtime_1 = require("../runtime");
const MetadataSchema_1 = require("./MetadataSchema");
const ReadCapacityResponse_1 = require("./ReadCapacityResponse");
/**
 * Check if a given object implements the ServerlessSpecResponse interface.
 */
function instanceOfServerlessSpecResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "cloud" in value;
    isInstance = isInstance && "region" in value;
    isInstance = isInstance && "readCapacity" in value;
    return isInstance;
}
function ServerlessSpecResponseFromJSON(json) {
    return ServerlessSpecResponseFromJSONTyped(json, false);
}
function ServerlessSpecResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'cloud': json['cloud'],
        'region': json['region'],
        'readCapacity': (0, ReadCapacityResponse_1.ReadCapacityResponseFromJSON)(json['read_capacity']),
        'sourceCollection': !(0, runtime_1.exists)(json, 'source_collection') ? undefined : json['source_collection'],
        'schema': !(0, runtime_1.exists)(json, 'schema') ? undefined : (0, MetadataSchema_1.MetadataSchemaFromJSON)(json['schema']),
    };
}
function ServerlessSpecResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'cloud': value.cloud,
        'region': value.region,
        'read_capacity': (0, ReadCapacityResponse_1.ReadCapacityResponseToJSON)(value.readCapacity),
        'source_collection': value.sourceCollection,
        'schema': (0, MetadataSchema_1.MetadataSchemaToJSON)(value.schema),
    };
}
