// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/ServerlessSpec.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfServerlessSpec = instanceOfServerlessSpec;
exports.ServerlessSpecFromJSON = ServerlessSpecFromJSON;
exports.ServerlessSpecFromJSONTyped = ServerlessSpecFromJSONTyped;
exports.ServerlessSpecToJSON = ServerlessSpecToJSON;
const runtime_1 = require("../runtime");
const MetadataSchema_1 = require("./MetadataSchema");
const ReadCapacity_1 = require("./ReadCapacity");
/**
 * Check if a given object implements the ServerlessSpec interface.
 */
function instanceOfServerlessSpec(value) {
    let isInstance = true;
    isInstance = isInstance && "cloud" in value;
    isInstance = isInstance && "region" in value;
    return isInstance;
}
function ServerlessSpecFromJSON(json) {
    return ServerlessSpecFromJSONTyped(json, false);
}
function ServerlessSpecFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'cloud': json['cloud'],
        'region': json['region'],
        'readCapacity': !(0, runtime_1.exists)(json, 'read_capacity') ? undefined : (0, ReadCapacity_1.ReadCapacityFromJSON)(json['read_capacity']),
        'sourceCollection': !(0, runtime_1.exists)(json, 'source_collection') ? undefined : json['source_collection'],
        'schema': !(0, runtime_1.exists)(json, 'schema') ? undefined : (0, MetadataSchema_1.MetadataSchemaFromJSON)(json['schema']),
    };
}
function ServerlessSpecToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'cloud': value.cloud,
        'region': value.region,
        'read_capacity': (0, ReadCapacity_1.ReadCapacityToJSON)(value.readCapacity),
        'source_collection': value.sourceCollection,
        'schema': (0, MetadataSchema_1.MetadataSchemaToJSON)(value.schema),
    };
}
