// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/Serverless1Serverless.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfServerless1Serverless = instanceOfServerless1Serverless;
exports.Serverless1ServerlessFromJSON = Serverless1ServerlessFromJSON;
exports.Serverless1ServerlessFromJSONTyped = Serverless1ServerlessFromJSONTyped;
exports.Serverless1ServerlessToJSON = Serverless1ServerlessToJSON;
const runtime_1 = require("../runtime");
const ReadCapacity_1 = require("./ReadCapacity");
/**
 * Check if a given object implements the Serverless1Serverless interface.
 */
function instanceOfServerless1Serverless(value) {
    let isInstance = true;
    return isInstance;
}
function Serverless1ServerlessFromJSON(json) {
    return Serverless1ServerlessFromJSONTyped(json, false);
}
function Serverless1ServerlessFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'readCapacity': !(0, runtime_1.exists)(json, 'read_capacity') ? undefined : (0, ReadCapacity_1.ReadCapacityFromJSON)(json['read_capacity']),
    };
}
function Serverless1ServerlessToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'read_capacity': (0, ReadCapacity_1.ReadCapacityToJSON)(value.readCapacity),
    };
}
