// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/Serverless.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfServerless = instanceOfServerless;
exports.ServerlessFromJSON = ServerlessFromJSON;
exports.ServerlessFromJSONTyped = ServerlessFromJSONTyped;
exports.ServerlessToJSON = ServerlessToJSON;
const ServerlessSpecResponse_1 = require("./ServerlessSpecResponse");
/**
 * Check if a given object implements the Serverless interface.
 */
function instanceOfServerless(value) {
    let isInstance = true;
    isInstance = isInstance && "serverless" in value;
    return isInstance;
}
function ServerlessFromJSON(json) {
    return ServerlessFromJSONTyped(json, false);
}
function ServerlessFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'serverless': (0, ServerlessSpecResponse_1.ServerlessSpecResponseFromJSON)(json['serverless']),
    };
}
function ServerlessToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'serverless': (0, ServerlessSpecResponse_1.ServerlessSpecResponseToJSON)(value.serverless),
    };
}
