// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/Serverless2.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfServerless2 = instanceOfServerless2;
exports.Serverless2FromJSON = Serverless2FromJSON;
exports.Serverless2FromJSONTyped = Serverless2FromJSONTyped;
exports.Serverless2ToJSON = Serverless2ToJSON;
const ServerlessSpec_1 = require("./ServerlessSpec");
/**
 * Check if a given object implements the Serverless2 interface.
 */
function instanceOfServerless2(value) {
    let isInstance = true;
    isInstance = isInstance && "serverless" in value;
    return isInstance;
}
function Serverless2FromJSON(json) {
    return Serverless2FromJSONTyped(json, false);
}
function Serverless2FromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'serverless': (0, ServerlessSpec_1.ServerlessSpecFromJSON)(json['serverless']),
    };
}
function Serverless2ToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'serverless': (0, ServerlessSpec_1.ServerlessSpecToJSON)(value.serverless),
    };
}
