// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_control/models/Serverless1.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Control Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfServerless1 = instanceOfServerless1;
exports.Serverless1FromJSON = Serverless1FromJSON;
exports.Serverless1FromJSONTyped = Serverless1FromJSONTyped;
exports.Serverless1ToJSON = Serverless1ToJSON;
const Serverless1Serverless_1 = require("./Serverless1Serverless");
/**
 * Check if a given object implements the Serverless1 interface.
 */
function instanceOfServerless1(value) {
    let isInstance = true;
    isInstance = isInstance && "serverless" in value;
    return isInstance;
}
function Serverless1FromJSON(json) {
    return Serverless1FromJSONTyped(json, false);
}
function Serverless1FromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'serverless': (0, Serverless1Serverless_1.Serverless1ServerlessFromJSON)(json['serverless']),
    };
}
function Serverless1ToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'serverless': (0, Serverless1Serverless_1.Serverless1ServerlessToJSON)(value.serverless),
    };
}
