// openai/resources/admin/organization/projects/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Users = exports.ServiceAccounts = exports.Roles = exports.RateLimits = exports.Projects = exports.Groups = exports.Certificates = exports.APIKeys = void 0;
var api_keys_1 = require("./api-keys.js");
Object.defineProperty(exports, "APIKeys", { enumerable: true, get: function () { return api_keys_1.APIKeys; } });
var certificates_1 = require("./certificates.js");
Object.defineProperty(exports, "Certificates", { enumerable: true, get: function () { return certificates_1.Certificates; } });
var index_1 = require("./groups/index.js");
Object.defineProperty(exports, "Groups", { enumerable: true, get: function () { return index_1.Groups; } });
var projects_1 = require("./projects.js");
Object.defineProperty(exports, "Projects", { enumerable: true, get: function () { return projects_1.Projects; } });
var rate_limits_1 = require("./rate-limits.js");
Object.defineProperty(exports, "RateLimits", { enumerable: true, get: function () { return rate_limits_1.RateLimits; } });
var roles_1 = require("./roles.js");
Object.defineProperty(exports, "Roles", { enumerable: true, get: function () { return roles_1.Roles; } });
var service_accounts_1 = require("./service-accounts.js");
Object.defineProperty(exports, "ServiceAccounts", { enumerable: true, get: function () { return service_accounts_1.ServiceAccounts; } });
var index_2 = require("./users/index.js");
Object.defineProperty(exports, "Users", { enumerable: true, get: function () { return index_2.Users; } });
