// @pinecone-database/pinecone/dist/assistant/control/createAssistant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAssistant = void 0;
const assistant_control_1 = require("../../pinecone-generated-ts-fetch/assistant_control");
const errors_1 = require("../../errors");
const createAssistant = (api) => {
    return async (options) => {
        validateCreateAssistantOptions(options);
        return await api.createAssistant({
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
            createAssistantRequest: {
                name: options.name,
                instructions: options?.instructions,
                metadata: options?.metadata,
                region: options?.region,
            },
        });
    };
};
exports.createAssistant = createAssistant;
const validateCreateAssistantOptions = (options) => {
    if (!options) {
        throw new errors_1.PineconeArgumentError('You must pass an object with required properties (`name`) to create an Assistant.');
    }
    if (options.region) {
        const normalizedRegion = options.region.toLowerCase();
        if (normalizedRegion !== 'us' && normalizedRegion !== 'eu') {
            throw new errors_1.PineconeArgumentError('Invalid region specified. Must be one of "us" or "eu"');
        }
        options.region = normalizedRegion;
    }
};
