// @pinecone-database/pinecone/dist/assistant/control/evaluate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.evaluate = void 0;
const assistant_evaluation_1 = require("../../pinecone-generated-ts-fetch/assistant_evaluation");
const errors_1 = require("../../errors");
const evaluate = (metricsApi) => {
    return async (options) => {
        if (!options) {
            throw new errors_1.PineconeArgumentError('You must pass an object with required properties (`question`, `answer`, `groundTruth`) to evaluate.');
        }
        if (options.question == '' ||
            options.answer == '' ||
            options.groundTruth == '') {
            throw new errors_1.PineconeArgumentError('Invalid input. Question, answer, and groundTruth must be non-empty strings.');
        }
        return await metricsApi.metricsAlignment({
            xPineconeApiVersion: assistant_evaluation_1.X_PINECONE_API_VERSION,
            alignmentRequest: {
                question: options.question,
                answer: options.answer,
                groundTruthAnswer: options.groundTruth,
            },
        });
    };
};
exports.evaluate = evaluate;
