// @pinecone-database/pinecone/dist/assistant/control/updateAssistant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAssistant = void 0;
const assistant_control_1 = require("../../pinecone-generated-ts-fetch/assistant_control");
const errors_1 = require("../../errors");
const updateAssistant = (api) => {
    return async (options) => {
        validateUpdateAssistantOptions(options);
        const updateAssistantRequest = {};
        if (options?.instructions) {
            updateAssistantRequest['instructions'] = options.instructions;
        }
        if (options?.metadata) {
            updateAssistantRequest['metadata'] = options.metadata;
        }
        return await api.updateAssistant({
            assistantName: options.name,
            updateAssistantRequest: updateAssistantRequest,
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
        });
    };
};
exports.updateAssistant = updateAssistant;
const validateUpdateAssistantOptions = (options) => {
    if (!options) {
        throw new errors_1.PineconeArgumentError('You must pass an object with at least one property to update an assistant.');
    }
    if (!options.name) {
        throw new errors_1.PineconeArgumentError('You must pass the name of an assistant to update.');
    }
};
