// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/CreateNamespaceRequestSchemaFieldsValue.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateNamespaceRequestSchemaFieldsValue = instanceOfCreateNamespaceRequestSchemaFieldsValue;
exports.CreateNamespaceRequestSchemaFieldsValueFromJSON = CreateNamespaceRequestSchemaFieldsValueFromJSON;
exports.CreateNamespaceRequestSchemaFieldsValueFromJSONTyped = CreateNamespaceRequestSchemaFieldsValueFromJSONTyped;
exports.CreateNamespaceRequestSchemaFieldsValueToJSON = CreateNamespaceRequestSchemaFieldsValueToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the CreateNamespaceRequestSchemaFieldsValue interface.
 */
function instanceOfCreateNamespaceRequestSchemaFieldsValue(value) {
    let isInstance = true;
    return isInstance;
}
function CreateNamespaceRequestSchemaFieldsValueFromJSON(json) {
    return CreateNamespaceRequestSchemaFieldsValueFromJSONTyped(json, false);
}
function CreateNamespaceRequestSchemaFieldsValueFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'filterable': !(0, runtime_1.exists)(json, 'filterable') ? undefined : json['filterable'],
    };
}
function CreateNamespaceRequestSchemaFieldsValueToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'filterable': value.filterable,
    };
}
