// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/CreateNamespaceRequestSchema.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateNamespaceRequestSchema = instanceOfCreateNamespaceRequestSchema;
exports.CreateNamespaceRequestSchemaFromJSON = CreateNamespaceRequestSchemaFromJSON;
exports.CreateNamespaceRequestSchemaFromJSONTyped = CreateNamespaceRequestSchemaFromJSONTyped;
exports.CreateNamespaceRequestSchemaToJSON = CreateNamespaceRequestSchemaToJSON;
const runtime_1 = require("../runtime");
const CreateNamespaceRequestSchemaFieldsValue_1 = require("./CreateNamespaceRequestSchemaFieldsValue");
/**
 * Check if a given object implements the CreateNamespaceRequestSchema interface.
 */
function instanceOfCreateNamespaceRequestSchema(value) {
    let isInstance = true;
    isInstance = isInstance && "fields" in value;
    return isInstance;
}
function CreateNamespaceRequestSchemaFromJSON(json) {
    return CreateNamespaceRequestSchemaFromJSONTyped(json, false);
}
function CreateNamespaceRequestSchemaFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fields': ((0, runtime_1.mapValues)(json['fields'], CreateNamespaceRequestSchemaFieldsValue_1.CreateNamespaceRequestSchemaFieldsValueFromJSON)),
    };
}
function CreateNamespaceRequestSchemaToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fields': ((0, runtime_1.mapValues)(value.fields, CreateNamespaceRequestSchemaFieldsValue_1.CreateNamespaceRequestSchemaFieldsValueToJSON)),
    };
}
