// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/CreateNamespaceRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateNamespaceRequest = instanceOfCreateNamespaceRequest;
exports.CreateNamespaceRequestFromJSON = CreateNamespaceRequestFromJSON;
exports.CreateNamespaceRequestFromJSONTyped = CreateNamespaceRequestFromJSONTyped;
exports.CreateNamespaceRequestToJSON = CreateNamespaceRequestToJSON;
const runtime_1 = require("../runtime");
const CreateNamespaceRequestSchema_1 = require("./CreateNamespaceRequestSchema");
/**
 * Check if a given object implements the CreateNamespaceRequest interface.
 */
function instanceOfCreateNamespaceRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "name" in value;
    return isInstance;
}
function CreateNamespaceRequestFromJSON(json) {
    return CreateNamespaceRequestFromJSONTyped(json, false);
}
function CreateNamespaceRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': json['name'],
        'schema': !(0, runtime_1.exists)(json, 'schema') ? undefined : (0, CreateNamespaceRequestSchema_1.CreateNamespaceRequestSchemaFromJSON)(json['schema']),
    };
}
function CreateNamespaceRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'schema': (0, CreateNamespaceRequestSchema_1.CreateNamespaceRequestSchemaToJSON)(value.schema),
    };
}
