// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/FetchByMetadataRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfFetchByMetadataRequest = instanceOfFetchByMetadataRequest;
exports.FetchByMetadataRequestFromJSON = FetchByMetadataRequestFromJSON;
exports.FetchByMetadataRequestFromJSONTyped = FetchByMetadataRequestFromJSONTyped;
exports.FetchByMetadataRequestToJSON = FetchByMetadataRequestToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the FetchByMetadataRequest interface.
 */
function instanceOfFetchByMetadataRequest(value) {
    let isInstance = true;
    return isInstance;
}
function FetchByMetadataRequestFromJSON(json) {
    return FetchByMetadataRequestFromJSONTyped(json, false);
}
function FetchByMetadataRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'namespace': !(0, runtime_1.exists)(json, 'namespace') ? undefined : json['namespace'],
        'filter': !(0, runtime_1.exists)(json, 'filter') ? undefined : json['filter'],
        'limit': !(0, runtime_1.exists)(json, 'limit') ? undefined : json['limit'],
        'paginationToken': !(0, runtime_1.exists)(json, 'paginationToken') ? undefined : json['paginationToken'],
    };
}
function FetchByMetadataRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'namespace': value.namespace,
        'filter': value.filter,
        'limit': value.limit,
        'paginationToken': value.paginationToken,
    };
}
