// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/ListResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfListResponse = instanceOfListResponse;
exports.ListResponseFromJSON = ListResponseFromJSON;
exports.ListResponseFromJSONTyped = ListResponseFromJSONTyped;
exports.ListResponseToJSON = ListResponseToJSON;
const runtime_1 = require("../runtime");
const ListItem_1 = require("./ListItem");
const Pagination_1 = require("./Pagination");
const Usage_1 = require("./Usage");
/**
 * Check if a given object implements the ListResponse interface.
 */
function instanceOfListResponse(value) {
    let isInstance = true;
    return isInstance;
}
function ListResponseFromJSON(json) {
    return ListResponseFromJSONTyped(json, false);
}
function ListResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'vectors': !(0, runtime_1.exists)(json, 'vectors') ? undefined : (json['vectors'].map(ListItem_1.ListItemFromJSON)),
        'pagination': !(0, runtime_1.exists)(json, 'pagination') ? undefined : (0, Pagination_1.PaginationFromJSON)(json['pagination']),
        'namespace': !(0, runtime_1.exists)(json, 'namespace') ? undefined : json['namespace'],
        'usage': !(0, runtime_1.exists)(json, 'usage') ? undefined : (0, Usage_1.UsageFromJSON)(json['usage']),
    };
}
function ListResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'vectors': value.vectors === undefined ? undefined : (value.vectors.map(ListItem_1.ListItemToJSON)),
        'pagination': (0, Pagination_1.PaginationToJSON)(value.pagination),
        'namespace': value.namespace,
        'usage': (0, Usage_1.UsageToJSON)(value.usage),
    };
}
