// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/ListItem.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfListItem = instanceOfListItem;
exports.ListItemFromJSON = ListItemFromJSON;
exports.ListItemFromJSONTyped = ListItemFromJSONTyped;
exports.ListItemToJSON = ListItemToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ListItem interface.
 */
function instanceOfListItem(value) {
    let isInstance = true;
    return isInstance;
}
function ListItemFromJSON(json) {
    return ListItemFromJSONTyped(json, false);
}
function ListItemFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'id': !(0, runtime_1.exists)(json, 'id') ? undefined : json['id'],
    };
}
function ListItemToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'id': value.id,
    };
}
