// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/NamespaceDescriptionIndexedFields.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfNamespaceDescriptionIndexedFields = instanceOfNamespaceDescriptionIndexedFields;
exports.NamespaceDescriptionIndexedFieldsFromJSON = NamespaceDescriptionIndexedFieldsFromJSON;
exports.NamespaceDescriptionIndexedFieldsFromJSONTyped = NamespaceDescriptionIndexedFieldsFromJSONTyped;
exports.NamespaceDescriptionIndexedFieldsToJSON = NamespaceDescriptionIndexedFieldsToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the NamespaceDescriptionIndexedFields interface.
 */
function instanceOfNamespaceDescriptionIndexedFields(value) {
    let isInstance = true;
    return isInstance;
}
function NamespaceDescriptionIndexedFieldsFromJSON(json) {
    return NamespaceDescriptionIndexedFieldsFromJSONTyped(json, false);
}
function NamespaceDescriptionIndexedFieldsFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fields': !(0, runtime_1.exists)(json, 'fields') ? undefined : json['fields'],
    };
}
function NamespaceDescriptionIndexedFieldsToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fields': value.fields,
    };
}
