// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/NamespaceSummary.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfNamespaceSummary = instanceOfNamespaceSummary;
exports.NamespaceSummaryFromJSON = NamespaceSummaryFromJSON;
exports.NamespaceSummaryFromJSONTyped = NamespaceSummaryFromJSONTyped;
exports.NamespaceSummaryToJSON = NamespaceSummaryToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the NamespaceSummary interface.
 */
function instanceOfNamespaceSummary(value) {
    let isInstance = true;
    return isInstance;
}
function NamespaceSummaryFromJSON(json) {
    return NamespaceSummaryFromJSONTyped(json, false);
}
function NamespaceSummaryFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'vectorCount': !(0, runtime_1.exists)(json, 'vectorCount') ? undefined : json['vectorCount'],
    };
}
function NamespaceSummaryToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'vectorCount': value.vectorCount,
    };
}
