// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/NamespaceDescription.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfNamespaceDescription = instanceOfNamespaceDescription;
exports.NamespaceDescriptionFromJSON = NamespaceDescriptionFromJSON;
exports.NamespaceDescriptionFromJSONTyped = NamespaceDescriptionFromJSONTyped;
exports.NamespaceDescriptionToJSON = NamespaceDescriptionToJSON;
const runtime_1 = require("../runtime");
const CreateNamespaceRequestSchema_1 = require("./CreateNamespaceRequestSchema");
const NamespaceDescriptionIndexedFields_1 = require("./NamespaceDescriptionIndexedFields");
/**
 * Check if a given object implements the NamespaceDescription interface.
 */
function instanceOfNamespaceDescription(value) {
    let isInstance = true;
    return isInstance;
}
function NamespaceDescriptionFromJSON(json) {
    return NamespaceDescriptionFromJSONTyped(json, false);
}
function NamespaceDescriptionFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': !(0, runtime_1.exists)(json, 'name') ? undefined : json['name'],
        'recordCount': !(0, runtime_1.exists)(json, 'record_count') ? undefined : json['record_count'],
        'schema': !(0, runtime_1.exists)(json, 'schema') ? undefined : (0, CreateNamespaceRequestSchema_1.CreateNamespaceRequestSchemaFromJSON)(json['schema']),
        'indexedFields': !(0, runtime_1.exists)(json, 'indexed_fields') ? undefined : (0, NamespaceDescriptionIndexedFields_1.NamespaceDescriptionIndexedFieldsFromJSON)(json['indexed_fields']),
    };
}
function NamespaceDescriptionToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'record_count': value.recordCount,
        'schema': (0, CreateNamespaceRequestSchema_1.CreateNamespaceRequestSchemaToJSON)(value.schema),
        'indexed_fields': (0, NamespaceDescriptionIndexedFields_1.NamespaceDescriptionIndexedFieldsToJSON)(value.indexedFields),
    };
}
