// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/RpcStatus.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfRpcStatus = instanceOfRpcStatus;
exports.RpcStatusFromJSON = RpcStatusFromJSON;
exports.RpcStatusFromJSONTyped = RpcStatusFromJSONTyped;
exports.RpcStatusToJSON = RpcStatusToJSON;
const runtime_1 = require("../runtime");
const ProtobufAny_1 = require("./ProtobufAny");
/**
 * Check if a given object implements the RpcStatus interface.
 */
function instanceOfRpcStatus(value) {
    let isInstance = true;
    return isInstance;
}
function RpcStatusFromJSON(json) {
    return RpcStatusFromJSONTyped(json, false);
}
function RpcStatusFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'code': !(0, runtime_1.exists)(json, 'code') ? undefined : json['code'],
        'message': !(0, runtime_1.exists)(json, 'message') ? undefined : json['message'],
        'details': !(0, runtime_1.exists)(json, 'details') ? undefined : (json['details'].map(ProtobufAny_1.ProtobufAnyFromJSON)),
    };
}
function RpcStatusToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'code': value.code,
        'message': value.message,
        'details': value.details === undefined ? undefined : (value.details.map(ProtobufAny_1.ProtobufAnyToJSON)),
    };
}
