// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/ScoredVector.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfScoredVector = instanceOfScoredVector;
exports.ScoredVectorFromJSON = ScoredVectorFromJSON;
exports.ScoredVectorFromJSONTyped = ScoredVectorFromJSONTyped;
exports.ScoredVectorToJSON = ScoredVectorToJSON;
const runtime_1 = require("../runtime");
const SparseValues_1 = require("./SparseValues");
/**
 * Check if a given object implements the ScoredVector interface.
 */
function instanceOfScoredVector(value) {
    let isInstance = true;
    isInstance = isInstance && "id" in value;
    return isInstance;
}
function ScoredVectorFromJSON(json) {
    return ScoredVectorFromJSONTyped(json, false);
}
function ScoredVectorFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'id': json['id'],
        'score': !(0, runtime_1.exists)(json, 'score') ? undefined : json['score'],
        'values': !(0, runtime_1.exists)(json, 'values') ? undefined : json['values'],
        'sparseValues': !(0, runtime_1.exists)(json, 'sparseValues') ? undefined : (0, SparseValues_1.SparseValuesFromJSON)(json['sparseValues']),
        'metadata': !(0, runtime_1.exists)(json, 'metadata') ? undefined : json['metadata'],
    };
}
function ScoredVectorToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'id': value.id,
        'score': value.score,
        'values': value.values,
        'sparseValues': (0, SparseValues_1.SparseValuesToJSON)(value.sparseValues),
        'metadata': value.metadata,
    };
}
