// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/SearchRecordsResponseResult.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchRecordsResponseResult = instanceOfSearchRecordsResponseResult;
exports.SearchRecordsResponseResultFromJSON = SearchRecordsResponseResultFromJSON;
exports.SearchRecordsResponseResultFromJSONTyped = SearchRecordsResponseResultFromJSONTyped;
exports.SearchRecordsResponseResultToJSON = SearchRecordsResponseResultToJSON;
const Hit_1 = require("./Hit");
/**
 * Check if a given object implements the SearchRecordsResponseResult interface.
 */
function instanceOfSearchRecordsResponseResult(value) {
    let isInstance = true;
    isInstance = isInstance && "hits" in value;
    return isInstance;
}
function SearchRecordsResponseResultFromJSON(json) {
    return SearchRecordsResponseResultFromJSONTyped(json, false);
}
function SearchRecordsResponseResultFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'hits': (json['hits'].map(Hit_1.HitFromJSON)),
    };
}
function SearchRecordsResponseResultToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'hits': (value.hits.map(Hit_1.HitToJSON)),
    };
}
