// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/SearchRecordsResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchRecordsResponse = instanceOfSearchRecordsResponse;
exports.SearchRecordsResponseFromJSON = SearchRecordsResponseFromJSON;
exports.SearchRecordsResponseFromJSONTyped = SearchRecordsResponseFromJSONTyped;
exports.SearchRecordsResponseToJSON = SearchRecordsResponseToJSON;
const SearchRecordsResponseResult_1 = require("./SearchRecordsResponseResult");
const SearchUsage_1 = require("./SearchUsage");
/**
 * Check if a given object implements the SearchRecordsResponse interface.
 */
function instanceOfSearchRecordsResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "result" in value;
    isInstance = isInstance && "usage" in value;
    return isInstance;
}
function SearchRecordsResponseFromJSON(json) {
    return SearchRecordsResponseFromJSONTyped(json, false);
}
function SearchRecordsResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'result': (0, SearchRecordsResponseResult_1.SearchRecordsResponseResultFromJSON)(json['result']),
        'usage': (0, SearchUsage_1.SearchUsageFromJSON)(json['usage']),
    };
}
function SearchRecordsResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'result': (0, SearchRecordsResponseResult_1.SearchRecordsResponseResultToJSON)(value.result),
        'usage': (0, SearchUsage_1.SearchUsageToJSON)(value.usage),
    };
}
