// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/SearchMatchTerms.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchMatchTerms = instanceOfSearchMatchTerms;
exports.SearchMatchTermsFromJSON = SearchMatchTermsFromJSON;
exports.SearchMatchTermsFromJSONTyped = SearchMatchTermsFromJSONTyped;
exports.SearchMatchTermsToJSON = SearchMatchTermsToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the SearchMatchTerms interface.
 */
function instanceOfSearchMatchTerms(value) {
    let isInstance = true;
    return isInstance;
}
function SearchMatchTermsFromJSON(json) {
    return SearchMatchTermsFromJSONTyped(json, false);
}
function SearchMatchTermsFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'strategy': !(0, runtime_1.exists)(json, 'strategy') ? undefined : json['strategy'],
        'terms': !(0, runtime_1.exists)(json, 'terms') ? undefined : json['terms'],
    };
}
function SearchMatchTermsToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'strategy': value.strategy,
        'terms': value.terms,
    };
}
