// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/SearchUsage.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchUsage = instanceOfSearchUsage;
exports.SearchUsageFromJSON = SearchUsageFromJSON;
exports.SearchUsageFromJSONTyped = SearchUsageFromJSONTyped;
exports.SearchUsageToJSON = SearchUsageToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the SearchUsage interface.
 */
function instanceOfSearchUsage(value) {
    let isInstance = true;
    isInstance = isInstance && "readUnits" in value;
    return isInstance;
}
function SearchUsageFromJSON(json) {
    return SearchUsageFromJSONTyped(json, false);
}
function SearchUsageFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'readUnits': json['read_units'],
        'embedTotalTokens': !(0, runtime_1.exists)(json, 'embed_total_tokens') ? undefined : json['embed_total_tokens'],
        'rerankUnits': !(0, runtime_1.exists)(json, 'rerank_units') ? undefined : json['rerank_units'],
    };
}
function SearchUsageToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'read_units': value.readUnits,
        'embed_total_tokens': value.embedTotalTokens,
        'rerank_units': value.rerankUnits,
    };
}
