// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/SearchRecordsRequestQuery.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchRecordsRequestQuery = instanceOfSearchRecordsRequestQuery;
exports.SearchRecordsRequestQueryFromJSON = SearchRecordsRequestQueryFromJSON;
exports.SearchRecordsRequestQueryFromJSONTyped = SearchRecordsRequestQueryFromJSONTyped;
exports.SearchRecordsRequestQueryToJSON = SearchRecordsRequestQueryToJSON;
const runtime_1 = require("../runtime");
const SearchMatchTerms_1 = require("./SearchMatchTerms");
const SearchRecordsVector_1 = require("./SearchRecordsVector");
/**
 * Check if a given object implements the SearchRecordsRequestQuery interface.
 */
function instanceOfSearchRecordsRequestQuery(value) {
    let isInstance = true;
    isInstance = isInstance && "topK" in value;
    return isInstance;
}
function SearchRecordsRequestQueryFromJSON(json) {
    return SearchRecordsRequestQueryFromJSONTyped(json, false);
}
function SearchRecordsRequestQueryFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'topK': json['top_k'],
        'filter': !(0, runtime_1.exists)(json, 'filter') ? undefined : json['filter'],
        'inputs': !(0, runtime_1.exists)(json, 'inputs') ? undefined : json['inputs'],
        'vector': !(0, runtime_1.exists)(json, 'vector') ? undefined : (0, SearchRecordsVector_1.SearchRecordsVectorFromJSON)(json['vector']),
        'id': !(0, runtime_1.exists)(json, 'id') ? undefined : json['id'],
        'matchTerms': !(0, runtime_1.exists)(json, 'match_terms') ? undefined : (0, SearchMatchTerms_1.SearchMatchTermsFromJSON)(json['match_terms']),
    };
}
function SearchRecordsRequestQueryToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'top_k': value.topK,
        'filter': value.filter,
        'inputs': value.inputs,
        'vector': (0, SearchRecordsVector_1.SearchRecordsVectorToJSON)(value.vector),
        'id': value.id,
        'match_terms': (0, SearchMatchTerms_1.SearchMatchTermsToJSON)(value.matchTerms),
    };
}
