// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/SearchRecordsRequestRerank.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchRecordsRequestRerank = instanceOfSearchRecordsRequestRerank;
exports.SearchRecordsRequestRerankFromJSON = SearchRecordsRequestRerankFromJSON;
exports.SearchRecordsRequestRerankFromJSONTyped = SearchRecordsRequestRerankFromJSONTyped;
exports.SearchRecordsRequestRerankToJSON = SearchRecordsRequestRerankToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the SearchRecordsRequestRerank interface.
 */
function instanceOfSearchRecordsRequestRerank(value) {
    let isInstance = true;
    isInstance = isInstance && "model" in value;
    isInstance = isInstance && "rankFields" in value;
    return isInstance;
}
function SearchRecordsRequestRerankFromJSON(json) {
    return SearchRecordsRequestRerankFromJSONTyped(json, false);
}
function SearchRecordsRequestRerankFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'model': json['model'],
        'rankFields': json['rank_fields'],
        'topN': !(0, runtime_1.exists)(json, 'top_n') ? undefined : json['top_n'],
        'parameters': !(0, runtime_1.exists)(json, 'parameters') ? undefined : json['parameters'],
        'query': !(0, runtime_1.exists)(json, 'query') ? undefined : json['query'],
    };
}
function SearchRecordsRequestRerankToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'model': value.model,
        'rank_fields': value.rankFields,
        'top_n': value.topN,
        'parameters': value.parameters,
        'query': value.query,
    };
}
