// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/UpdateResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfUpdateResponse = instanceOfUpdateResponse;
exports.UpdateResponseFromJSON = UpdateResponseFromJSON;
exports.UpdateResponseFromJSONTyped = UpdateResponseFromJSONTyped;
exports.UpdateResponseToJSON = UpdateResponseToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the UpdateResponse interface.
 */
function instanceOfUpdateResponse(value) {
    let isInstance = true;
    return isInstance;
}
function UpdateResponseFromJSON(json) {
    return UpdateResponseFromJSONTyped(json, false);
}
function UpdateResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'matchedRecords': !(0, runtime_1.exists)(json, 'matchedRecords') ? undefined : json['matchedRecords'],
    };
}
function UpdateResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'matchedRecords': value.matchedRecords,
    };
}
