// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/UpdateRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfUpdateRequest = instanceOfUpdateRequest;
exports.UpdateRequestFromJSON = UpdateRequestFromJSON;
exports.UpdateRequestFromJSONTyped = UpdateRequestFromJSONTyped;
exports.UpdateRequestToJSON = UpdateRequestToJSON;
const runtime_1 = require("../runtime");
const SparseValues_1 = require("./SparseValues");
/**
 * Check if a given object implements the UpdateRequest interface.
 */
function instanceOfUpdateRequest(value) {
    let isInstance = true;
    return isInstance;
}
function UpdateRequestFromJSON(json) {
    return UpdateRequestFromJSONTyped(json, false);
}
function UpdateRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'id': !(0, runtime_1.exists)(json, 'id') ? undefined : json['id'],
        'values': !(0, runtime_1.exists)(json, 'values') ? undefined : json['values'],
        'sparseValues': !(0, runtime_1.exists)(json, 'sparseValues') ? undefined : (0, SparseValues_1.SparseValuesFromJSON)(json['sparseValues']),
        'setMetadata': !(0, runtime_1.exists)(json, 'setMetadata') ? undefined : json['setMetadata'],
        'namespace': !(0, runtime_1.exists)(json, 'namespace') ? undefined : json['namespace'],
        'filter': !(0, runtime_1.exists)(json, 'filter') ? undefined : json['filter'],
        'dryRun': !(0, runtime_1.exists)(json, 'dryRun') ? undefined : json['dryRun'],
    };
}
function UpdateRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'id': value.id,
        'values': value.values,
        'sparseValues': (0, SparseValues_1.SparseValuesToJSON)(value.sparseValues),
        'setMetadata': value.setMetadata,
        'namespace': value.namespace,
        'filter': value.filter,
        'dryRun': value.dryRun,
    };
}
