// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/db_data/models/Usage.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Data Plane API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfUsage = instanceOfUsage;
exports.UsageFromJSON = UsageFromJSON;
exports.UsageFromJSONTyped = UsageFromJSONTyped;
exports.UsageToJSON = UsageToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the Usage interface.
 */
function instanceOfUsage(value) {
    let isInstance = true;
    return isInstance;
}
function UsageFromJSON(json) {
    return UsageFromJSONTyped(json, false);
}
function UsageFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'readUnits': !(0, runtime_1.exists)(json, 'readUnits') ? undefined : json['readUnits'],
    };
}
function UsageToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'readUnits': value.readUnits,
    };
}
