// weaviate-client/dist/node/esm/collections/tenants/index.js
import { WeaviateUnexpectedStatusCodeError } from '../../errors.js';
import { TenantsCreator, TenantsDeleter, TenantsUpdater } from '../../schema/index.js';
import { Deserialize } from '../deserialize/index.js';
import { Serialize } from '../serialize/index.js';
const parseValueOrValueArray = (value) => (Array.isArray(value) ? value : [value]);
const parseStringOrTenant = (tenant) => typeof tenant === 'string' ? tenant : tenant.name;
const parseTenantREST = (tenant) => ({
    name: tenant.name,
    activityStatus: Deserialize.activityStatusREST(tenant.activityStatus),
});
const tenants = (connection, collection, dbVersionSupport) => {
    const get = (names) => connection
        .tenants(collection)
        .then((builder) => builder.withGet({ names }))
        .then(Deserialize.tenantsGet);
    const update = async (tenants) => {
        const out = [];
        for await (const res of Serialize.tenants(parseValueOrValueArray(tenants), Serialize.tenantUpdate).map((tenants) => new TenantsUpdater(connection, collection, tenants).do().then((res) => res.map(parseTenantREST)))) {
            out.push(...res);
        }
        return out;
    };
    return {
        create: (tenants) => new TenantsCreator(connection, collection, parseValueOrValueArray(tenants).map(Serialize.tenantCreate))
            .do()
            .then((res) => res.map(parseTenantREST)),
        get: () => get(),
        getByNames: (tenants) => get(tenants.map(parseStringOrTenant)),
        getByName: async (tenant) => {
            const tenantName = parseStringOrTenant(tenant);
            if (await dbVersionSupport.supportsTenantGetRESTMethod().then((check) => !check.supports)) {
                return get([tenantName]).then((tenants) => tenants[tenantName] ?? null);
            }
            return connection
                .get(`/schema/${collection}/tenants/${tenantName}`)
                .then(parseTenantREST)
                .catch((err) => {
                if (err instanceof WeaviateUnexpectedStatusCodeError && err.code === 404) {
                    return null;
                }
                throw err;
            });
        },
        remove: (tenants) => new TenantsDeleter(connection, collection, parseValueOrValueArray(tenants).map(parseStringOrTenant)).do(),
        update,
        activate: (tenant) => {
            return update(parseValueOrValueArray(tenant).map((tenant) => ({
                name: parseStringOrTenant(tenant),
                activityStatus: 'ACTIVE',
            })));
        },
        deactivate: (tenant) => {
            return update(parseValueOrValueArray(tenant).map((tenant) => ({
                name: parseStringOrTenant(tenant),
                activityStatus: 'INACTIVE',
            })));
        },
        offload: (tenant) => {
            return update(parseValueOrValueArray(tenant).map((tenant) => ({
                name: parseStringOrTenant(tenant),
                activityStatus: 'OFFLOADED',
            })));
        },
    };
};
export default tenants;
