// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/DenseEmbedding.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfDenseEmbedding = instanceOfDenseEmbedding;
exports.DenseEmbeddingFromJSON = DenseEmbeddingFromJSON;
exports.DenseEmbeddingFromJSONTyped = DenseEmbeddingFromJSONTyped;
exports.DenseEmbeddingToJSON = DenseEmbeddingToJSON;
/**
 * Check if a given object implements the DenseEmbedding interface.
 */
function instanceOfDenseEmbedding(value) {
    let isInstance = true;
    isInstance = isInstance && "values" in value;
    isInstance = isInstance && "vectorType" in value;
    return isInstance;
}
function DenseEmbeddingFromJSON(json) {
    return DenseEmbeddingFromJSONTyped(json, false);
}
function DenseEmbeddingFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'values': json['values'],
        'vectorType': json['vector_type'],
    };
}
function DenseEmbeddingToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'values': value.values,
        'vector_type': value.vectorType,
    };
}
