// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/EmbedRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfEmbedRequest = instanceOfEmbedRequest;
exports.EmbedRequestFromJSON = EmbedRequestFromJSON;
exports.EmbedRequestFromJSONTyped = EmbedRequestFromJSONTyped;
exports.EmbedRequestToJSON = EmbedRequestToJSON;
const runtime_1 = require("../runtime");
const EmbedRequestInputsInner_1 = require("./EmbedRequestInputsInner");
/**
 * Check if a given object implements the EmbedRequest interface.
 */
function instanceOfEmbedRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "model" in value;
    isInstance = isInstance && "inputs" in value;
    return isInstance;
}
function EmbedRequestFromJSON(json) {
    return EmbedRequestFromJSONTyped(json, false);
}
function EmbedRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'model': json['model'],
        'parameters': !(0, runtime_1.exists)(json, 'parameters') ? undefined : json['parameters'],
        'inputs': (json['inputs'].map(EmbedRequestInputsInner_1.EmbedRequestInputsInnerFromJSON)),
    };
}
function EmbedRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'model': value.model,
        'parameters': value.parameters,
        'inputs': (value.inputs.map(EmbedRequestInputsInner_1.EmbedRequestInputsInnerToJSON)),
    };
}
