// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/EmbeddingsList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfEmbeddingsList = instanceOfEmbeddingsList;
exports.EmbeddingsListFromJSON = EmbeddingsListFromJSON;
exports.EmbeddingsListFromJSONTyped = EmbeddingsListFromJSONTyped;
exports.EmbeddingsListToJSON = EmbeddingsListToJSON;
const Embedding_1 = require("./Embedding");
const EmbeddingsListUsage_1 = require("./EmbeddingsListUsage");
/**
 * Check if a given object implements the EmbeddingsList interface.
 */
function instanceOfEmbeddingsList(value) {
    let isInstance = true;
    isInstance = isInstance && "model" in value;
    isInstance = isInstance && "vectorType" in value;
    isInstance = isInstance && "data" in value;
    isInstance = isInstance && "usage" in value;
    return isInstance;
}
function EmbeddingsListFromJSON(json) {
    return EmbeddingsListFromJSONTyped(json, false);
}
function EmbeddingsListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'model': json['model'],
        'vectorType': json['vector_type'],
        'data': (json['data'].map(Embedding_1.EmbeddingFromJSON)),
        'usage': (0, EmbeddingsListUsage_1.EmbeddingsListUsageFromJSON)(json['usage']),
    };
}
function EmbeddingsListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'model': value.model,
        'vector_type': value.vectorType,
        'data': (value.data.map(Embedding_1.EmbeddingToJSON)),
        'usage': (0, EmbeddingsListUsage_1.EmbeddingsListUsageToJSON)(value.usage),
    };
}
