// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/EmbeddingsListUsage.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfEmbeddingsListUsage = instanceOfEmbeddingsListUsage;
exports.EmbeddingsListUsageFromJSON = EmbeddingsListUsageFromJSON;
exports.EmbeddingsListUsageFromJSONTyped = EmbeddingsListUsageFromJSONTyped;
exports.EmbeddingsListUsageToJSON = EmbeddingsListUsageToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the EmbeddingsListUsage interface.
 */
function instanceOfEmbeddingsListUsage(value) {
    let isInstance = true;
    return isInstance;
}
function EmbeddingsListUsageFromJSON(json) {
    return EmbeddingsListUsageFromJSONTyped(json, false);
}
function EmbeddingsListUsageFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'totalTokens': !(0, runtime_1.exists)(json, 'total_tokens') ? undefined : json['total_tokens'],
    };
}
function EmbeddingsListUsageToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'total_tokens': value.totalTokens,
    };
}
