// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/EmbedRequestInputsInner.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfEmbedRequestInputsInner = instanceOfEmbedRequestInputsInner;
exports.EmbedRequestInputsInnerFromJSON = EmbedRequestInputsInnerFromJSON;
exports.EmbedRequestInputsInnerFromJSONTyped = EmbedRequestInputsInnerFromJSONTyped;
exports.EmbedRequestInputsInnerToJSON = EmbedRequestInputsInnerToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the EmbedRequestInputsInner interface.
 */
function instanceOfEmbedRequestInputsInner(value) {
    let isInstance = true;
    return isInstance;
}
function EmbedRequestInputsInnerFromJSON(json) {
    return EmbedRequestInputsInnerFromJSONTyped(json, false);
}
function EmbedRequestInputsInnerFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'text': !(0, runtime_1.exists)(json, 'text') ? undefined : json['text'],
    };
}
function EmbedRequestInputsInnerToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'text': value.text,
    };
}
