// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/ModelInfoList.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfModelInfoList = instanceOfModelInfoList;
exports.ModelInfoListFromJSON = ModelInfoListFromJSON;
exports.ModelInfoListFromJSONTyped = ModelInfoListFromJSONTyped;
exports.ModelInfoListToJSON = ModelInfoListToJSON;
const runtime_1 = require("../runtime");
const ModelInfo_1 = require("./ModelInfo");
/**
 * Check if a given object implements the ModelInfoList interface.
 */
function instanceOfModelInfoList(value) {
    let isInstance = true;
    return isInstance;
}
function ModelInfoListFromJSON(json) {
    return ModelInfoListFromJSONTyped(json, false);
}
function ModelInfoListFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'models': !(0, runtime_1.exists)(json, 'models') ? undefined : (json['models'].map(ModelInfo_1.ModelInfoFromJSON)),
    };
}
function ModelInfoListToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'models': value.models === undefined ? undefined : (value.models.map(ModelInfo_1.ModelInfoToJSON)),
    };
}
