// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/ModelInfoSupportedParameterAllowedValuesInner.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfModelInfoSupportedParameterAllowedValuesInner = instanceOfModelInfoSupportedParameterAllowedValuesInner;
exports.ModelInfoSupportedParameterAllowedValuesInnerFromJSON = ModelInfoSupportedParameterAllowedValuesInnerFromJSON;
exports.ModelInfoSupportedParameterAllowedValuesInnerFromJSONTyped = ModelInfoSupportedParameterAllowedValuesInnerFromJSONTyped;
exports.ModelInfoSupportedParameterAllowedValuesInnerToJSON = ModelInfoSupportedParameterAllowedValuesInnerToJSON;
/**
 * Check if a given object implements the ModelInfoSupportedParameterAllowedValuesInner interface.
 */
function instanceOfModelInfoSupportedParameterAllowedValuesInner(value) {
    let isInstance = true;
    return isInstance;
}
function ModelInfoSupportedParameterAllowedValuesInnerFromJSON(json) {
    return ModelInfoSupportedParameterAllowedValuesInnerFromJSONTyped(json, false);
}
function ModelInfoSupportedParameterAllowedValuesInnerFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ModelInfoSupportedParameterAllowedValuesInnerToJSON(value) {
    return value;
}
