// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/ModelInfoSupportedParameterDefault.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfModelInfoSupportedParameterDefault = instanceOfModelInfoSupportedParameterDefault;
exports.ModelInfoSupportedParameterDefaultFromJSON = ModelInfoSupportedParameterDefaultFromJSON;
exports.ModelInfoSupportedParameterDefaultFromJSONTyped = ModelInfoSupportedParameterDefaultFromJSONTyped;
exports.ModelInfoSupportedParameterDefaultToJSON = ModelInfoSupportedParameterDefaultToJSON;
/**
 * Check if a given object implements the ModelInfoSupportedParameterDefault interface.
 */
function instanceOfModelInfoSupportedParameterDefault(value) {
    let isInstance = true;
    return isInstance;
}
function ModelInfoSupportedParameterDefaultFromJSON(json) {
    return ModelInfoSupportedParameterDefaultFromJSONTyped(json, false);
}
function ModelInfoSupportedParameterDefaultFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
function ModelInfoSupportedParameterDefaultToJSON(value) {
    return value;
}
