// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/RankedDocument.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfRankedDocument = instanceOfRankedDocument;
exports.RankedDocumentFromJSON = RankedDocumentFromJSON;
exports.RankedDocumentFromJSONTyped = RankedDocumentFromJSONTyped;
exports.RankedDocumentToJSON = RankedDocumentToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the RankedDocument interface.
 */
function instanceOfRankedDocument(value) {
    let isInstance = true;
    isInstance = isInstance && "index" in value;
    isInstance = isInstance && "score" in value;
    return isInstance;
}
function RankedDocumentFromJSON(json) {
    return RankedDocumentFromJSONTyped(json, false);
}
function RankedDocumentFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'index': json['index'],
        'score': json['score'],
        'document': !(0, runtime_1.exists)(json, 'document') ? undefined : json['document'],
    };
}
function RankedDocumentToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'index': value.index,
        'score': value.score,
        'document': value.document,
    };
}
