// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/inference/models/RerankResultUsage.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Inference API
 * Pinecone is a vector database that makes it easy to search and retrieve billions of high-dimensional vectors.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfRerankResultUsage = instanceOfRerankResultUsage;
exports.RerankResultUsageFromJSON = RerankResultUsageFromJSON;
exports.RerankResultUsageFromJSONTyped = RerankResultUsageFromJSONTyped;
exports.RerankResultUsageToJSON = RerankResultUsageToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the RerankResultUsage interface.
 */
function instanceOfRerankResultUsage(value) {
    let isInstance = true;
    return isInstance;
}
function RerankResultUsageFromJSON(json) {
    return RerankResultUsageFromJSONTyped(json, false);
}
function RerankResultUsageFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'rerankUnits': !(0, runtime_1.exists)(json, 'rerank_units') ? undefined : json['rerank_units'],
    };
}
function RerankResultUsageToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'rerank_units': value.rerankUnits,
    };
}
