// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_control/models/CreateAssistantRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Control Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge  from millions of documents at scale. This API supports creating and managing assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfCreateAssistantRequest = instanceOfCreateAssistantRequest;
exports.CreateAssistantRequestFromJSON = CreateAssistantRequestFromJSON;
exports.CreateAssistantRequestFromJSONTyped = CreateAssistantRequestFromJSONTyped;
exports.CreateAssistantRequestToJSON = CreateAssistantRequestToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the CreateAssistantRequest interface.
 */
function instanceOfCreateAssistantRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "name" in value;
    return isInstance;
}
function CreateAssistantRequestFromJSON(json) {
    return CreateAssistantRequestFromJSONTyped(json, false);
}
function CreateAssistantRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'name': json['name'],
        'instructions': !(0, runtime_1.exists)(json, 'instructions') ? undefined : json['instructions'],
        'metadata': !(0, runtime_1.exists)(json, 'metadata') ? undefined : json['metadata'],
        'region': !(0, runtime_1.exists)(json, 'region') ? undefined : json['region'],
    };
}
function CreateAssistantRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'name': value.name,
        'instructions': value.instructions,
        'metadata': value.metadata,
        'region': value.region,
    };
}
