// weaviate-client/dist/node/cjs/tokenize/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_js_1 = require("./util.js");
const tokenize = (connection, dbVersionSupport) => {
    const parseTextAnalyzerConfig = (config) => {
        if (config == undefined)
            return undefined;
        const out = { stopwordPreset: config.stopwordPreset ? String(config.stopwordPreset) : undefined };
        if (typeof config?.asciiFold === 'boolean') {
            return { ...out, asciiFold: config?.asciiFold };
        }
        else if (typeof config?.asciiFold === 'object') {
            return {
                ...out,
                asciiFold: true,
                asciiFoldIgnore: config?.asciiFold.ignore,
            };
        }
        return out;
    };
    return {
        text: (text, tokenization, opts) => {
            return dbVersionSupport
                .supportsTokenize()
                .then(({ supports, message }) => (supports ? Promise.resolve() : Promise.reject(new Error(message))))
                .then(() => connection
                .postReturn('/tokenize', {
                text,
                tokenization,
                analyzerConfig: parseTextAnalyzerConfig(opts?.analyzerConfig),
                stopwordPresets: opts?.stopwordPresets,
            })
                .then(util_js_1.parseResult));
        },
        forProperty: (collection, property, text) => dbVersionSupport
            .supportsTokenize()
            .then(({ supports, message }) => (supports ? Promise.resolve() : Promise.reject(new Error(message))))
            .then(() => connection.postReturn(`/schema/${collection}/properties/${property}/tokenize`, { text }))
            .then(util_js_1.parseResult),
    };
};
exports.default = tokenize;
