// weaviate-client/dist/node/esm/collections/configure/parsing.js
export class QuantizerGuards {
    static isPQCreate(config) {
        return config?.type === 'pq';
    }
    static isPQUpdate(config) {
        return config?.type === 'pq';
    }
    static isBQCreate(config) {
        return config?.type === 'bq';
    }
    static isBQUpdate(config) {
        return config?.type === 'bq';
    }
    static isSQCreate(config) {
        return config?.type === 'sq';
    }
    static isSQUpdate(config) {
        return config?.type === 'sq';
    }
    static isRQCreate(config) {
        return config?.type === 'rq';
    }
    static isRQUpdate(config) {
        return config?.type === 'rq';
    }
    static isUncompressedCreate(config) {
        return config?.type === 'none';
    }
}
export class VectorIndexGuards {
    static isHNSW(config) {
        return config?.type === 'hnsw';
    }
    static isFlat(config) {
        return config?.type === 'flat';
    }
    static isHFresh(config) {
        return config?.type === 'hfresh';
    }
    static isDynamic(config) {
        return config?.type === 'dynamic';
    }
}
export class MultiVectorEncodingGuards {
    static isMuvera(config) {
        return config?.type === 'muvera';
    }
}
export function parseWithDefault(value, defaultValue) {
    return value !== undefined ? value : defaultValue;
}
