// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/AlignmentRequest.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfAlignmentRequest = instanceOfAlignmentRequest;
exports.AlignmentRequestFromJSON = AlignmentRequestFromJSON;
exports.AlignmentRequestFromJSONTyped = AlignmentRequestFromJSONTyped;
exports.AlignmentRequestToJSON = AlignmentRequestToJSON;
/**
 * Check if a given object implements the AlignmentRequest interface.
 */
function instanceOfAlignmentRequest(value) {
    let isInstance = true;
    isInstance = isInstance && "question" in value;
    isInstance = isInstance && "answer" in value;
    isInstance = isInstance && "groundTruthAnswer" in value;
    return isInstance;
}
function AlignmentRequestFromJSON(json) {
    return AlignmentRequestFromJSONTyped(json, false);
}
function AlignmentRequestFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'question': json['question'],
        'answer': json['answer'],
        'groundTruthAnswer': json['ground_truth_answer'],
    };
}
function AlignmentRequestToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'question': value.question,
        'answer': value.answer,
        'ground_truth_answer': value.groundTruthAnswer,
    };
}
