// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/BasicErrorResponse.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfBasicErrorResponse = instanceOfBasicErrorResponse;
exports.BasicErrorResponseFromJSON = BasicErrorResponseFromJSON;
exports.BasicErrorResponseFromJSONTyped = BasicErrorResponseFromJSONTyped;
exports.BasicErrorResponseToJSON = BasicErrorResponseToJSON;
/**
 * Check if a given object implements the BasicErrorResponse interface.
 */
function instanceOfBasicErrorResponse(value) {
    let isInstance = true;
    isInstance = isInstance && "message" in value;
    return isInstance;
}
function BasicErrorResponseFromJSON(json) {
    return BasicErrorResponseFromJSONTyped(json, false);
}
function BasicErrorResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'message': json['message'],
    };
}
function BasicErrorResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'message': value.message,
    };
}
