// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/EvaluatedFact.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfEvaluatedFact = instanceOfEvaluatedFact;
exports.EvaluatedFactFromJSON = EvaluatedFactFromJSON;
exports.EvaluatedFactFromJSONTyped = EvaluatedFactFromJSONTyped;
exports.EvaluatedFactToJSON = EvaluatedFactToJSON;
const Fact_1 = require("./Fact");
/**
 * Check if a given object implements the EvaluatedFact interface.
 */
function instanceOfEvaluatedFact(value) {
    let isInstance = true;
    isInstance = isInstance && "fact" in value;
    isInstance = isInstance && "entailment" in value;
    return isInstance;
}
function EvaluatedFactFromJSON(json) {
    return EvaluatedFactFromJSONTyped(json, false);
}
function EvaluatedFactFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'fact': (0, Fact_1.FactFromJSON)(json['fact']),
        'entailment': json['entailment'],
    };
}
function EvaluatedFactToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'fact': (0, Fact_1.FactToJSON)(value.fact),
        'entailment': value.entailment,
    };
}
