// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/Fact.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfFact = instanceOfFact;
exports.FactFromJSON = FactFromJSON;
exports.FactFromJSONTyped = FactFromJSONTyped;
exports.FactToJSON = FactToJSON;
/**
 * Check if a given object implements the Fact interface.
 */
function instanceOfFact(value) {
    let isInstance = true;
    isInstance = isInstance && "content" in value;
    return isInstance;
}
function FactFromJSON(json) {
    return FactFromJSONTyped(json, false);
}
function FactFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'content': json['content'],
    };
}
function FactToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'content': value.content,
    };
}
