// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/Metrics.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMetrics = instanceOfMetrics;
exports.MetricsFromJSON = MetricsFromJSON;
exports.MetricsFromJSONTyped = MetricsFromJSONTyped;
exports.MetricsToJSON = MetricsToJSON;
/**
 * Check if a given object implements the Metrics interface.
 */
function instanceOfMetrics(value) {
    let isInstance = true;
    isInstance = isInstance && "correctness" in value;
    isInstance = isInstance && "completeness" in value;
    isInstance = isInstance && "alignment" in value;
    return isInstance;
}
function MetricsFromJSON(json) {
    return MetricsFromJSONTyped(json, false);
}
function MetricsFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'correctness': json['correctness'],
        'completeness': json['completeness'],
        'alignment': json['alignment'],
    };
}
function MetricsToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'correctness': value.correctness,
        'completeness': value.completeness,
        'alignment': value.alignment,
    };
}
