// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/Reasoning.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReasoning = instanceOfReasoning;
exports.ReasoningFromJSON = ReasoningFromJSON;
exports.ReasoningFromJSONTyped = ReasoningFromJSONTyped;
exports.ReasoningToJSON = ReasoningToJSON;
const EvaluatedFact_1 = require("./EvaluatedFact");
/**
 * Check if a given object implements the Reasoning interface.
 */
function instanceOfReasoning(value) {
    let isInstance = true;
    isInstance = isInstance && "evaluatedFacts" in value;
    return isInstance;
}
function ReasoningFromJSON(json) {
    return ReasoningFromJSONTyped(json, false);
}
function ReasoningFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'evaluatedFacts': (json['evaluated_facts'].map(EvaluatedFact_1.EvaluatedFactFromJSON)),
    };
}
function ReasoningToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'evaluated_facts': (value.evaluatedFacts.map(EvaluatedFact_1.EvaluatedFactToJSON)),
    };
}
