// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_evaluation/models/TokenCounts.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Evaluation API
 * Provides endpoints for evaluating RAG systems using various metrics.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfTokenCounts = instanceOfTokenCounts;
exports.TokenCountsFromJSON = TokenCountsFromJSON;
exports.TokenCountsFromJSONTyped = TokenCountsFromJSONTyped;
exports.TokenCountsToJSON = TokenCountsToJSON;
/**
 * Check if a given object implements the TokenCounts interface.
 */
function instanceOfTokenCounts(value) {
    let isInstance = true;
    isInstance = isInstance && "promptTokens" in value;
    isInstance = isInstance && "completionTokens" in value;
    isInstance = isInstance && "totalTokens" in value;
    return isInstance;
}
function TokenCountsFromJSON(json) {
    return TokenCountsFromJSONTyped(json, false);
}
function TokenCountsFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'promptTokens': json['prompt_tokens'],
        'completionTokens': json['completion_tokens'],
        'totalTokens': json['total_tokens'],
    };
}
function TokenCountsToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'prompt_tokens': value.promptTokens,
        'completion_tokens': value.completionTokens,
        'total_tokens': value.totalTokens,
    };
}
