// @pinecone-database/pinecone/dist/assistant/control/__tests__/assistantOperations.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assistant_control_1 = require("../../../pinecone-generated-ts-fetch/assistant_control");
const describeAssistant_1 = require("../describeAssistant");
const deleteAssistant_1 = require("../deleteAssistant");
const listAssistants_1 = require("../listAssistants");
const setupManageAssistantsApi = () => {
    const fakeGetAssistant = jest.fn().mockImplementation(() => Promise.resolve({
        name: 'test-assistant',
        status: 'Ready',
        instructions: 'test instructions',
        metadata: { key: 'value' },
        host: 'https://test-host.com',
        createdAt: new Date(),
        updatedAt: new Date(),
    }));
    const fakeDeleteAssistant = jest.fn().mockImplementation(() => Promise.resolve());
    const fakeListAssistants = jest.fn().mockImplementation(() => Promise.resolve({
        assistants: [
            {
                name: 'assistant-1',
                status: 'Ready',
                host: 'https://test-host.com',
                createdAt: new Date(),
                updatedAt: new Date(),
            },
            {
                name: 'assistant-2',
                status: 'Initializing',
                host: 'https://test-host.com',
                createdAt: new Date(),
                updatedAt: new Date(),
            },
        ],
    }));
    const MAP = {
        getAssistant: fakeGetAssistant,
        deleteAssistant: fakeDeleteAssistant,
        listAssistants: fakeListAssistants,
    };
    return MAP;
};
describe('describeAssistant', () => {
    let manageAssistantsApi;
    beforeEach(() => {
        manageAssistantsApi = setupManageAssistantsApi();
    });
    test('throws error when name is not provided', async () => {
        await expect((0, describeAssistant_1.describeAssistant)(manageAssistantsApi)('')).rejects.toThrow('You must pass the name of an assistant to update.');
    });
    test('calls getAssistant with correct parameters', async () => {
        await (0, describeAssistant_1.describeAssistant)(manageAssistantsApi)('test-assistant');
        expect(manageAssistantsApi.getAssistant).toHaveBeenCalledWith({
            assistantName: 'test-assistant',
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
        });
    });
});
describe('deleteAssistant', () => {
    let manageAssistantsApi;
    beforeEach(() => {
        manageAssistantsApi = setupManageAssistantsApi();
    });
    test('throws error when name is not provided', async () => {
        await expect((0, deleteAssistant_1.deleteAssistant)(manageAssistantsApi)('')).rejects.toThrow('You must pass the name of an assistant to delete.');
    });
    test('calls deleteAssistant with correct parameters', async () => {
        await (0, deleteAssistant_1.deleteAssistant)(manageAssistantsApi)('test-assistant');
        expect(manageAssistantsApi.deleteAssistant).toHaveBeenCalledWith({
            assistantName: 'test-assistant',
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
        });
    });
    test('resolves successfully when delete completes', async () => {
        await expect((0, deleteAssistant_1.deleteAssistant)(manageAssistantsApi)('test-assistant')).resolves.toBeUndefined();
    });
});
describe('listAssistants', () => {
    let manageAssistantsApi;
    beforeEach(() => {
        manageAssistantsApi = setupManageAssistantsApi();
    });
    test('calls listAssistants with correct parameters', async () => {
        await (0, listAssistants_1.listAssistants)(manageAssistantsApi)();
        expect(manageAssistantsApi.listAssistants).toHaveBeenCalledWith({
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
        });
    });
});
