// @pinecone-database/pinecone/dist/assistant/control/__tests__/evaluate.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const evaluate_1 = require("../evaluate");
const assistant_evaluation_1 = require("../../../pinecone-generated-ts-fetch/assistant_evaluation");
const setupMetricsApi = () => {
    const fakeMetricsAlignment = jest
        .fn()
        .mockImplementation(() => Promise.resolve({}));
    const MAP = {
        metricsAlignment: fakeMetricsAlignment,
    };
    return MAP;
};
describe('AssistantCtrlPlane', () => {
    let metricsApi;
    beforeAll(() => {
        metricsApi = setupMetricsApi();
    });
    describe('evaluate', () => {
        test('throws error when options is null or undefined', async () => {
            await expect(
            // @ts-expect-error - invalid options
            (0, evaluate_1.evaluate)(metricsApi)(null)).rejects.toThrow('You must pass an object with required properties (`question`, `answer`, `groundTruth`) to evaluate.');
            await expect(
            // @ts-expect-error - invalid options
            (0, evaluate_1.evaluate)(metricsApi)(undefined)).rejects.toThrow('You must pass an object with required properties (`question`, `answer`, `groundTruth`) to evaluate.');
        });
        test('throws error when empty strings are provided', async () => {
            const emptyRequests = [
                {
                    question: '',
                    answer: 'test-answer',
                    groundTruth: 'test-ground-truth',
                },
                {
                    question: 'test-question',
                    answer: '',
                    groundTruth: 'test-ground-truth',
                },
                {
                    question: 'test-question',
                    answer: 'test-answer',
                    groundTruth: '',
                },
            ];
            for (const request of emptyRequests) {
                await expect((0, evaluate_1.evaluate)(metricsApi)(request)).rejects.toThrow('Invalid input. Question, answer, and groundTruth must be non-empty strings.');
            }
        });
        test('calls metricsAlignment with correct parameters', async () => {
            const request = {
                question: 'What is the capital of France?',
                answer: 'Paris',
                groundTruth: 'The capital of France is Paris.',
            };
            await (0, evaluate_1.evaluate)(metricsApi)(request);
            expect(metricsApi.metricsAlignment).toHaveBeenCalledWith({
                xPineconeApiVersion: assistant_evaluation_1.X_PINECONE_API_VERSION,
                alignmentRequest: {
                    question: request.question,
                    answer: request.answer,
                    groundTruthAnswer: request.groundTruth,
                },
            });
        });
    });
});
