// @pinecone-database/pinecone/dist/assistant/control/__tests__/updateAssistant.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assistant_control_1 = require("../../../pinecone-generated-ts-fetch/assistant_control");
const updateAssistant_1 = require("../updateAssistant");
const setupManageAssistantsApi = () => {
    const fakeUpdateAssistant = jest.fn().mockImplementation(() => Promise.resolve({
        assistantName: 'test-assistant',
        instructions: 'updated instructions',
        metadata: { key: 'value' },
    }));
    const MAP = {
        updateAssistant: fakeUpdateAssistant,
    };
    return MAP;
};
describe('updateAssistant', () => {
    let manageAssistantsApi;
    beforeEach(() => {
        manageAssistantsApi = setupManageAssistantsApi();
    });
    test('throws error when name is not provided', async () => {
        await expect((0, updateAssistant_1.updateAssistant)(manageAssistantsApi)({
            name: '',
            instructions: 'new instructions',
        })).rejects.toThrow('You must pass the name of an assistant to update.');
    });
    test('calls updateAssistant with correct parameters', async () => {
        const options = {
            name: 'test-assistant',
            instructions: 'new instructions',
            metadata: { key: 'value' },
        };
        await (0, updateAssistant_1.updateAssistant)(manageAssistantsApi)(options);
        expect(manageAssistantsApi.updateAssistant).toHaveBeenCalledWith({
            assistantName: 'test-assistant',
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
            updateAssistantRequest: {
                instructions: 'new instructions',
                metadata: { key: 'value' },
            },
        });
    });
    test('updates only instructions when metadata not provided', async () => {
        const options = {
            name: 'test-assistant',
            instructions: 'new instructions',
        };
        await (0, updateAssistant_1.updateAssistant)(manageAssistantsApi)(options);
        expect(manageAssistantsApi.updateAssistant).toHaveBeenCalledWith({
            assistantName: 'test-assistant',
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
            updateAssistantRequest: {
                instructions: 'new instructions',
                metadata: undefined,
            },
        });
    });
    test('updates only metadata when instructions not provided', async () => {
        const options = {
            name: 'test-assistant',
            metadata: { key: 'new-value' },
        };
        await (0, updateAssistant_1.updateAssistant)(manageAssistantsApi)(options);
        expect(manageAssistantsApi.updateAssistant).toHaveBeenCalledWith({
            assistantName: 'test-assistant',
            xPineconeApiVersion: assistant_control_1.X_PINECONE_API_VERSION,
            updateAssistantRequest: {
                instructions: undefined,
                metadata: { key: 'new-value' },
            },
        });
    });
});
