// @pinecone-database/pinecone/dist/control/__tests__/configureIndex.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configureIndex_1 = require("../configureIndex");
const api_version_1 = require("../../pinecone-generated-ts-fetch/db_control/api_version");
describe('configureIndex', () => {
    const podIndexModel = {
        name: 'pod-index',
        dimension: 5,
        metric: 'cosine',
        host: 'https://index-host.com',
        vectorType: 'dense',
        spec: {
            pod: {
                environment: 'us-east1-gcp',
                replicas: 4,
                shards: 1,
                pods: 4,
                podType: 'p2.x2',
            },
        },
        status: {
            ready: true,
            state: 'Ready',
        },
        tags: {
            example: 'tag',
        },
        deletionProtection: 'disabled',
    };
    const serverlessIndexModel = {
        name: 'serverless-index',
        dimension: 5,
        metric: 'cosine',
        host: 'https://serverless-host.com',
        vectorType: 'dense',
        spec: {
            serverless: {
                cloud: 'aws',
                region: 'us-east-1',
                readCapacity: {
                    mode: 'OnDemand',
                    status: {
                        state: 'Ready',
                    },
                },
            },
        },
        status: {
            ready: true,
            state: 'Ready',
        },
        deletionProtection: 'enabled',
    };
    const byocIndexModel = {
        name: 'byoc-index',
        dimension: 5,
        metric: 'cosine',
        host: 'https://byoc-host.com',
        vectorType: 'dense',
        spec: {
            byoc: {
                environment: 'my-byoc-env',
                readCapacity: {
                    mode: 'OnDemand',
                    status: {
                        state: 'Ready',
                    },
                },
            },
        },
        status: {
            ready: true,
            state: 'Ready',
        },
        deletionProtection: 'disabled',
    };
    test('calls the openapi configure endpoint with pod index configuration', async () => {
        const fakeDescribe = jest.fn().mockResolvedValue(podIndexModel);
        const fakeConfigure = jest.fn().mockResolvedValue(podIndexModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: fakeConfigure,
        };
        const returned = await (0, configureIndex_1.configureIndex)(IOA)({
            name: 'pod-index',
            podReplicas: 4,
            podType: 'p2.x2',
            deletionProtection: 'disabled',
            tags: {
                example: 'tag',
            },
        });
        expect(returned).toBe(podIndexModel);
        expect(IOA.describeIndex).toHaveBeenCalledWith({
            indexName: 'pod-index',
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
        expect(IOA.configureIndex).toHaveBeenCalledWith({
            indexName: 'pod-index',
            configureIndexRequest: {
                spec: { pod: { replicas: 4, podType: 'p2.x2' } },
                deletionProtection: 'disabled',
                tags: {
                    example: 'tag',
                },
            },
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
    });
    test('calls the openapi configure endpoint with serverless index configuration', async () => {
        const fakeDescribe = jest.fn().mockResolvedValue(serverlessIndexModel);
        const fakeConfigure = jest
            .fn()
            .mockResolvedValue(serverlessIndexModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: fakeConfigure,
        };
        const returned = await (0, configureIndex_1.configureIndex)(IOA)({
            name: 'serverless-index',
            readCapacity: {
                mode: 'Dedicated',
                nodeType: 't1',
                manual: {
                    replicas: 2,
                    shards: 3,
                },
            },
        });
        expect(returned).toBe(serverlessIndexModel);
        expect(IOA.describeIndex).toHaveBeenCalledWith({
            indexName: 'serverless-index',
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
        expect(IOA.configureIndex).toHaveBeenCalledWith({
            indexName: 'serverless-index',
            configureIndexRequest: {
                spec: {
                    serverless: {
                        readCapacity: {
                            mode: 'Dedicated',
                            dedicated: {
                                nodeType: 't1',
                                scaling: 'Manual',
                                manual: {
                                    replicas: 2,
                                    shards: 3,
                                },
                            },
                        },
                    },
                },
            },
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
    });
    test('calls the openapi configure endpoint with BYOC index configuration', async () => {
        const fakeDescribe = jest.fn().mockResolvedValue(byocIndexModel);
        const fakeConfigure = jest.fn().mockResolvedValue(byocIndexModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: fakeConfigure,
        };
        const returned = await (0, configureIndex_1.configureIndex)(IOA)({
            name: 'byoc-index',
            readCapacity: {
                mode: 'OnDemand',
            },
        });
        expect(returned).toBe(byocIndexModel);
        expect(IOA.describeIndex).toHaveBeenCalledWith({
            indexName: 'byoc-index',
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
        expect(IOA.configureIndex).toHaveBeenCalledWith({
            indexName: 'byoc-index',
            configureIndexRequest: {
                spec: {
                    byoc: {
                        readCapacity: {
                            mode: 'OnDemand',
                        },
                    },
                },
            },
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
    });
    test('throws error when trying to configure readCapacity on pod index', async () => {
        const fakeDescribe = jest.fn().mockResolvedValue(podIndexModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: jest.fn(),
        };
        await expect((0, configureIndex_1.configureIndex)(IOA)({
            name: 'pod-index',
            readCapacity: { mode: 'OnDemand' },
        })).rejects.toThrow('Cannot configure readCapacity on a pod index; readCapacity is only supported for serverless and BYOC indexes.');
        expect(IOA.describeIndex).toHaveBeenCalled();
        expect(IOA.configureIndex).not.toHaveBeenCalled();
    });
    test('throws error when trying to configure podReplicas on serverless index', async () => {
        const fakeDescribe = jest.fn().mockResolvedValue(serverlessIndexModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: jest.fn(),
        };
        await expect((0, configureIndex_1.configureIndex)(IOA)({
            name: 'serverless-index',
            podReplicas: 4,
        })).rejects.toThrow('Cannot configure podReplicas or podType on a serverless index; these parameters are only supported for pod indexes.');
        expect(IOA.describeIndex).toHaveBeenCalled();
        expect(IOA.configureIndex).not.toHaveBeenCalled();
    });
    test('throws error when trying to configure podType on BYOC index', async () => {
        const fakeDescribe = jest.fn().mockResolvedValue(byocIndexModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: jest.fn(),
        };
        await expect((0, configureIndex_1.configureIndex)(IOA)({
            name: 'byoc-index',
            podType: 'p1.x1',
        })).rejects.toThrow('Cannot configure podReplicas or podType on a byoc index; these parameters are only supported for pod indexes.');
        expect(IOA.describeIndex).toHaveBeenCalled();
        expect(IOA.configureIndex).not.toHaveBeenCalled();
    });
    test('skips describeIndex when only non-spec params are updated', async () => {
        const fakeConfigure = jest.fn().mockResolvedValue(podIndexModel);
        const IOA = {
            describeIndex: jest.fn(),
            configureIndex: fakeConfigure,
        };
        const returned = await (0, configureIndex_1.configureIndex)(IOA)({
            name: 'pod-index',
            deletionProtection: 'enabled',
            tags: { env: 'prod' },
        });
        expect(returned).toBe(podIndexModel);
        // describeIndex must NOT be called — no spec params were supplied
        expect(IOA.describeIndex).not.toHaveBeenCalled();
        expect(IOA.configureIndex).toHaveBeenCalledWith({
            indexName: 'pod-index',
            configureIndexRequest: {
                spec: undefined,
                deletionProtection: 'enabled',
                tags: { env: 'prod' },
                embed: undefined,
            },
            xPineconeApiVersion: api_version_1.X_PINECONE_API_VERSION,
        });
    });
    test('throws error when spec type cannot be determined and spec params are supplied', async () => {
        // Return an IndexModel whose spec has no recognisable type
        const unknownSpecModel = {
            ...podIndexModel,
            spec: {},
        };
        const fakeDescribe = jest.fn().mockResolvedValue(unknownSpecModel);
        const IOA = {
            describeIndex: fakeDescribe,
            configureIndex: jest.fn(),
        };
        await expect((0, configureIndex_1.configureIndex)(IOA)({
            name: 'pod-index',
            podReplicas: 2,
        })).rejects.toThrow('Could not determine the index spec type. Verify the index exists and try again.');
        expect(IOA.describeIndex).toHaveBeenCalled();
        expect(IOA.configureIndex).not.toHaveBeenCalled();
    });
});
describe('getIndexSpecType', () => {
    // IndexModelSpecFromJSON merges BYOC, PodBased, Serverless FromJSON results,
    // so spec can have pod/serverless/byoc all present with some undefined.
    // getIndexSpecType must classify by defined value, not key presence.
    test('returns serverless when only serverless has a defined value (merged FromJSON shape)', () => {
        const spec = {
            pod: undefined,
            serverless: {
                cloud: 'aws',
                region: 'us-east-1',
                readCapacity: { mode: 'OnDemand', status: { state: 'Ready' } },
            },
            byoc: undefined,
        };
        expect((0, configureIndex_1.getIndexSpecType)(spec)).toBe('serverless');
    });
    test('returns pod when only pod has a defined value (merged FromJSON shape)', () => {
        const spec = {
            pod: {
                environment: 'us-east1-gcp',
                replicas: 4,
                shards: 1,
                pods: 4,
                podType: 'p2.x2',
            },
            serverless: undefined,
            byoc: undefined,
        };
        expect((0, configureIndex_1.getIndexSpecType)(spec)).toBe('pod');
    });
    test('returns byoc when only byoc has a defined value (merged FromJSON shape)', () => {
        const spec = {
            pod: undefined,
            serverless: undefined,
            byoc: {
                environment: 'my-env',
                readCapacity: { mode: 'OnDemand', status: { state: 'Ready' } },
            },
        };
        expect((0, configureIndex_1.getIndexSpecType)(spec)).toBe('byoc');
    });
    test('returns unknown for null or non-object', () => {
        expect((0, configureIndex_1.getIndexSpecType)(null)).toBe('unknown');
        expect((0, configureIndex_1.getIndexSpecType)(undefined)).toBe('unknown');
    });
});
