// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ChatModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfChatModel = instanceOfChatModel;
exports.ChatModelFromJSON = ChatModelFromJSON;
exports.ChatModelFromJSONTyped = ChatModelFromJSONTyped;
exports.ChatModelToJSON = ChatModelToJSON;
const runtime_1 = require("../runtime");
const CitationModel_1 = require("./CitationModel");
const MessageModel_1 = require("./MessageModel");
const UsageModel_1 = require("./UsageModel");
/**
 * Check if a given object implements the ChatModel interface.
 */
function instanceOfChatModel(value) {
    let isInstance = true;
    return isInstance;
}
function ChatModelFromJSON(json) {
    return ChatModelFromJSONTyped(json, false);
}
function ChatModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'id': !(0, runtime_1.exists)(json, 'id') ? undefined : json['id'],
        'finishReason': !(0, runtime_1.exists)(json, 'finish_reason') ? undefined : json['finish_reason'],
        'message': !(0, runtime_1.exists)(json, 'message') ? undefined : (0, MessageModel_1.MessageModelFromJSON)(json['message']),
        'model': !(0, runtime_1.exists)(json, 'model') ? undefined : json['model'],
        'citations': !(0, runtime_1.exists)(json, 'citations') ? undefined : (json['citations'].map(CitationModel_1.CitationModelFromJSON)),
        'usage': !(0, runtime_1.exists)(json, 'usage') ? undefined : (0, UsageModel_1.UsageModelFromJSON)(json['usage']),
    };
}
function ChatModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'id': value.id,
        'finish_reason': value.finishReason,
        'message': (0, MessageModel_1.MessageModelToJSON)(value.message),
        'model': value.model,
        'citations': value.citations === undefined ? undefined : (value.citations.map(CitationModel_1.CitationModelToJSON)),
        'usage': (0, UsageModel_1.UsageModelToJSON)(value.usage),
    };
}
