// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ChoiceModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfChoiceModel = instanceOfChoiceModel;
exports.ChoiceModelFromJSON = ChoiceModelFromJSON;
exports.ChoiceModelFromJSONTyped = ChoiceModelFromJSONTyped;
exports.ChoiceModelToJSON = ChoiceModelToJSON;
const runtime_1 = require("../runtime");
const MessageModel_1 = require("./MessageModel");
/**
 * Check if a given object implements the ChoiceModel interface.
 */
function instanceOfChoiceModel(value) {
    let isInstance = true;
    return isInstance;
}
function ChoiceModelFromJSON(json) {
    return ChoiceModelFromJSONTyped(json, false);
}
function ChoiceModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'finishReason': !(0, runtime_1.exists)(json, 'finish_reason') ? undefined : json['finish_reason'],
        'index': !(0, runtime_1.exists)(json, 'index') ? undefined : json['index'],
        'message': !(0, runtime_1.exists)(json, 'message') ? undefined : (0, MessageModel_1.MessageModelFromJSON)(json['message']),
    };
}
function ChoiceModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'finish_reason': value.finishReason,
        'index': value.index,
        'message': (0, MessageModel_1.MessageModelToJSON)(value.message),
    };
}
