// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ChoiceChunkModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfChoiceChunkModel = instanceOfChoiceChunkModel;
exports.ChoiceChunkModelFromJSON = ChoiceChunkModelFromJSON;
exports.ChoiceChunkModelFromJSONTyped = ChoiceChunkModelFromJSONTyped;
exports.ChoiceChunkModelToJSON = ChoiceChunkModelToJSON;
const runtime_1 = require("../runtime");
const ChoiceChunkModelDelta_1 = require("./ChoiceChunkModelDelta");
/**
 * Check if a given object implements the ChoiceChunkModel interface.
 */
function instanceOfChoiceChunkModel(value) {
    let isInstance = true;
    return isInstance;
}
function ChoiceChunkModelFromJSON(json) {
    return ChoiceChunkModelFromJSONTyped(json, false);
}
function ChoiceChunkModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'finishReason': !(0, runtime_1.exists)(json, 'finish_reason') ? undefined : json['finish_reason'],
        'index': !(0, runtime_1.exists)(json, 'index') ? undefined : json['index'],
        'delta': !(0, runtime_1.exists)(json, 'delta') ? undefined : (0, ChoiceChunkModelDelta_1.ChoiceChunkModelDeltaFromJSON)(json['delta']),
    };
}
function ChoiceChunkModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'finish_reason': value.finishReason,
        'index': value.index,
        'delta': (0, ChoiceChunkModelDelta_1.ChoiceChunkModelDeltaToJSON)(value.delta),
    };
}
