// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ChoiceChunkModelDelta.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfChoiceChunkModelDelta = instanceOfChoiceChunkModelDelta;
exports.ChoiceChunkModelDeltaFromJSON = ChoiceChunkModelDeltaFromJSON;
exports.ChoiceChunkModelDeltaFromJSONTyped = ChoiceChunkModelDeltaFromJSONTyped;
exports.ChoiceChunkModelDeltaToJSON = ChoiceChunkModelDeltaToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ChoiceChunkModelDelta interface.
 */
function instanceOfChoiceChunkModelDelta(value) {
    let isInstance = true;
    return isInstance;
}
function ChoiceChunkModelDeltaFromJSON(json) {
    return ChoiceChunkModelDeltaFromJSONTyped(json, false);
}
function ChoiceChunkModelDeltaFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'role': !(0, runtime_1.exists)(json, 'role') ? undefined : json['role'],
        'content': !(0, runtime_1.exists)(json, 'content') ? undefined : json['content'],
    };
}
function ChoiceChunkModelDeltaToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'role': value.role,
        'content': value.content,
    };
}
