// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/DocxReferenceModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfDocxReferenceModel = instanceOfDocxReferenceModel;
exports.DocxReferenceModelFromJSON = DocxReferenceModelFromJSON;
exports.DocxReferenceModelFromJSONTyped = DocxReferenceModelFromJSONTyped;
exports.DocxReferenceModelToJSON = DocxReferenceModelToJSON;
const AssistantFileModel_1 = require("./AssistantFileModel");
/**
 * Check if a given object implements the DocxReferenceModel interface.
 */
function instanceOfDocxReferenceModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "file" in value;
    isInstance = isInstance && "pages" in value;
    return isInstance;
}
function DocxReferenceModelFromJSON(json) {
    return DocxReferenceModelFromJSONTyped(json, false);
}
function DocxReferenceModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'file': (0, AssistantFileModel_1.AssistantFileModelFromJSON)(json['file']),
        'pages': json['pages'],
    };
}
function DocxReferenceModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'file': (0, AssistantFileModel_1.AssistantFileModelToJSON)(value.file),
        'pages': value.pages,
    };
}
