// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ErrorResponseError.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfErrorResponseError = instanceOfErrorResponseError;
exports.ErrorResponseErrorFromJSON = ErrorResponseErrorFromJSON;
exports.ErrorResponseErrorFromJSONTyped = ErrorResponseErrorFromJSONTyped;
exports.ErrorResponseErrorToJSON = ErrorResponseErrorToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the ErrorResponseError interface.
 */
function instanceOfErrorResponseError(value) {
    let isInstance = true;
    isInstance = isInstance && "code" in value;
    isInstance = isInstance && "message" in value;
    return isInstance;
}
function ErrorResponseErrorFromJSON(json) {
    return ErrorResponseErrorFromJSONTyped(json, false);
}
function ErrorResponseErrorFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'code': json['code'],
        'message': json['message'],
        'details': !(0, runtime_1.exists)(json, 'details') ? undefined : json['details'],
    };
}
function ErrorResponseErrorToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'code': value.code,
        'message': value.message,
        'details': value.details,
    };
}
