// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/HighlightModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfHighlightModel = instanceOfHighlightModel;
exports.HighlightModelFromJSON = HighlightModelFromJSON;
exports.HighlightModelFromJSONTyped = HighlightModelFromJSONTyped;
exports.HighlightModelToJSON = HighlightModelToJSON;
/**
 * Check if a given object implements the HighlightModel interface.
 */
function instanceOfHighlightModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "content" in value;
    return isInstance;
}
function HighlightModelFromJSON(json) {
    return HighlightModelFromJSONTyped(json, false);
}
function HighlightModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'content': json['content'],
    };
}
function HighlightModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'content': value.content,
    };
}
