// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ImageModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfImageModel = instanceOfImageModel;
exports.ImageModelFromJSON = ImageModelFromJSON;
exports.ImageModelFromJSONTyped = ImageModelFromJSONTyped;
exports.ImageModelToJSON = ImageModelToJSON;
/**
 * Check if a given object implements the ImageModel interface.
 */
function instanceOfImageModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "mimeType" in value;
    isInstance = isInstance && "data" in value;
    return isInstance;
}
function ImageModelFromJSON(json) {
    return ImageModelFromJSONTyped(json, false);
}
function ImageModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'mimeType': json['mime_type'],
        'data': json['data'],
    };
}
function ImageModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'mime_type': value.mimeType,
        'data': value.data,
    };
}
