// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ListFiles200Response.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfListFiles200Response = instanceOfListFiles200Response;
exports.ListFiles200ResponseFromJSON = ListFiles200ResponseFromJSON;
exports.ListFiles200ResponseFromJSONTyped = ListFiles200ResponseFromJSONTyped;
exports.ListFiles200ResponseToJSON = ListFiles200ResponseToJSON;
const runtime_1 = require("../runtime");
const AssistantFileModel_1 = require("./AssistantFileModel");
/**
 * Check if a given object implements the ListFiles200Response interface.
 */
function instanceOfListFiles200Response(value) {
    let isInstance = true;
    return isInstance;
}
function ListFiles200ResponseFromJSON(json) {
    return ListFiles200ResponseFromJSONTyped(json, false);
}
function ListFiles200ResponseFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'files': !(0, runtime_1.exists)(json, 'files') ? undefined : (json['files'].map(AssistantFileModel_1.AssistantFileModelFromJSON)),
    };
}
function ListFiles200ResponseToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'files': value.files === undefined ? undefined : (value.files.map(AssistantFileModel_1.AssistantFileModelToJSON)),
    };
}
