// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/MessageModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMessageModel = instanceOfMessageModel;
exports.MessageModelFromJSON = MessageModelFromJSON;
exports.MessageModelFromJSONTyped = MessageModelFromJSONTyped;
exports.MessageModelToJSON = MessageModelToJSON;
const runtime_1 = require("../runtime");
/**
 * Check if a given object implements the MessageModel interface.
 */
function instanceOfMessageModel(value) {
    let isInstance = true;
    return isInstance;
}
function MessageModelFromJSON(json) {
    return MessageModelFromJSONTyped(json, false);
}
function MessageModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'role': !(0, runtime_1.exists)(json, 'role') ? undefined : json['role'],
        'content': !(0, runtime_1.exists)(json, 'content') ? undefined : json['content'],
    };
}
function MessageModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'role': value.role,
        'content': value.content,
    };
}
