// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/MultiModalContentTextBlockModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMultiModalContentTextBlockModel = instanceOfMultiModalContentTextBlockModel;
exports.MultiModalContentTextBlockModelFromJSON = MultiModalContentTextBlockModelFromJSON;
exports.MultiModalContentTextBlockModelFromJSONTyped = MultiModalContentTextBlockModelFromJSONTyped;
exports.MultiModalContentTextBlockModelToJSON = MultiModalContentTextBlockModelToJSON;
/**
 * Check if a given object implements the MultiModalContentTextBlockModel interface.
 */
function instanceOfMultiModalContentTextBlockModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "text" in value;
    return isInstance;
}
function MultiModalContentTextBlockModelFromJSON(json) {
    return MultiModalContentTextBlockModelFromJSONTyped(json, false);
}
function MultiModalContentTextBlockModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'text': json['text'],
    };
}
function MultiModalContentTextBlockModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'text': value.text,
    };
}
