// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/MultiModalContentImageBlockModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMultiModalContentImageBlockModel = instanceOfMultiModalContentImageBlockModel;
exports.MultiModalContentImageBlockModelFromJSON = MultiModalContentImageBlockModelFromJSON;
exports.MultiModalContentImageBlockModelFromJSONTyped = MultiModalContentImageBlockModelFromJSONTyped;
exports.MultiModalContentImageBlockModelToJSON = MultiModalContentImageBlockModelToJSON;
const runtime_1 = require("../runtime");
const ImageModel_1 = require("./ImageModel");
/**
 * Check if a given object implements the MultiModalContentImageBlockModel interface.
 */
function instanceOfMultiModalContentImageBlockModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "caption" in value;
    return isInstance;
}
function MultiModalContentImageBlockModelFromJSON(json) {
    return MultiModalContentImageBlockModelFromJSONTyped(json, false);
}
function MultiModalContentImageBlockModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'caption': json['caption'],
        'image': !(0, runtime_1.exists)(json, 'image') ? undefined : (0, ImageModel_1.ImageModelFromJSON)(json['image']),
    };
}
function MultiModalContentImageBlockModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'caption': value.caption,
        'image': (0, ImageModel_1.ImageModelToJSON)(value.image),
    };
}
