// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/MultiModalSnippetModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfMultiModalSnippetModel = instanceOfMultiModalSnippetModel;
exports.MultiModalSnippetModelFromJSON = MultiModalSnippetModelFromJSON;
exports.MultiModalSnippetModelFromJSONTyped = MultiModalSnippetModelFromJSONTyped;
exports.MultiModalSnippetModelToJSON = MultiModalSnippetModelToJSON;
const MultiModalContentBlocksModel_1 = require("./MultiModalContentBlocksModel");
const TypedReferenceModel_1 = require("./TypedReferenceModel");
/**
 * Check if a given object implements the MultiModalSnippetModel interface.
 */
function instanceOfMultiModalSnippetModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "content" in value;
    isInstance = isInstance && "score" in value;
    isInstance = isInstance && "reference" in value;
    return isInstance;
}
function MultiModalSnippetModelFromJSON(json) {
    return MultiModalSnippetModelFromJSONTyped(json, false);
}
function MultiModalSnippetModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'content': (json['content'].map(MultiModalContentBlocksModel_1.MultiModalContentBlocksModelFromJSON)),
        'score': json['score'],
        'reference': (0, TypedReferenceModel_1.TypedReferenceModelFromJSON)(json['reference']),
    };
}
function MultiModalSnippetModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'content': (value.content.map(MultiModalContentBlocksModel_1.MultiModalContentBlocksModelToJSON)),
        'score': value.score,
        'reference': (0, TypedReferenceModel_1.TypedReferenceModelToJSON)(value.reference),
    };
}
