// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/ReferenceModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfReferenceModel = instanceOfReferenceModel;
exports.ReferenceModelFromJSON = ReferenceModelFromJSON;
exports.ReferenceModelFromJSONTyped = ReferenceModelFromJSONTyped;
exports.ReferenceModelToJSON = ReferenceModelToJSON;
const runtime_1 = require("../runtime");
const AssistantFileModel_1 = require("./AssistantFileModel");
const HighlightModel_1 = require("./HighlightModel");
/**
 * Check if a given object implements the ReferenceModel interface.
 */
function instanceOfReferenceModel(value) {
    let isInstance = true;
    return isInstance;
}
function ReferenceModelFromJSON(json) {
    return ReferenceModelFromJSONTyped(json, false);
}
function ReferenceModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'file': !(0, runtime_1.exists)(json, 'file') ? undefined : (0, AssistantFileModel_1.AssistantFileModelFromJSON)(json['file']),
        'pages': !(0, runtime_1.exists)(json, 'pages') ? undefined : json['pages'],
        'highlight': !(0, runtime_1.exists)(json, 'highlight') ? undefined : (0, HighlightModel_1.HighlightModelFromJSON)(json['highlight']),
    };
}
function ReferenceModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'file': (0, AssistantFileModel_1.AssistantFileModelToJSON)(value.file),
        'pages': value.pages,
        'highlight': (0, HighlightModel_1.HighlightModelToJSON)(value.highlight),
    };
}
