// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/SearchCompletions.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfSearchCompletions = instanceOfSearchCompletions;
exports.SearchCompletionsFromJSON = SearchCompletionsFromJSON;
exports.SearchCompletionsFromJSONTyped = SearchCompletionsFromJSONTyped;
exports.SearchCompletionsToJSON = SearchCompletionsToJSON;
const runtime_1 = require("../runtime");
const MessageModel_1 = require("./MessageModel");
/**
 * Check if a given object implements the SearchCompletions interface.
 */
function instanceOfSearchCompletions(value) {
    let isInstance = true;
    isInstance = isInstance && "messages" in value;
    return isInstance;
}
function SearchCompletionsFromJSON(json) {
    return SearchCompletionsFromJSONTyped(json, false);
}
function SearchCompletionsFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'messages': (json['messages'].map(MessageModel_1.MessageModelFromJSON)),
        'stream': !(0, runtime_1.exists)(json, 'stream') ? undefined : json['stream'],
        'model': !(0, runtime_1.exists)(json, 'model') ? undefined : json['model'],
        'temperature': !(0, runtime_1.exists)(json, 'temperature') ? undefined : json['temperature'],
        'filter': !(0, runtime_1.exists)(json, 'filter') ? undefined : json['filter'],
    };
}
function SearchCompletionsToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'messages': (value.messages.map(MessageModel_1.MessageModelToJSON)),
        'stream': value.stream,
        'model': value.model,
        'temperature': value.temperature,
        'filter': value.filter,
    };
}
