// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/StreamChatCompletionChunkModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfStreamChatCompletionChunkModel = instanceOfStreamChatCompletionChunkModel;
exports.StreamChatCompletionChunkModelFromJSON = StreamChatCompletionChunkModelFromJSON;
exports.StreamChatCompletionChunkModelFromJSONTyped = StreamChatCompletionChunkModelFromJSONTyped;
exports.StreamChatCompletionChunkModelToJSON = StreamChatCompletionChunkModelToJSON;
const runtime_1 = require("../runtime");
const ChoiceChunkModel_1 = require("./ChoiceChunkModel");
/**
 * Check if a given object implements the StreamChatCompletionChunkModel interface.
 */
function instanceOfStreamChatCompletionChunkModel(value) {
    let isInstance = true;
    return isInstance;
}
function StreamChatCompletionChunkModelFromJSON(json) {
    return StreamChatCompletionChunkModelFromJSONTyped(json, false);
}
function StreamChatCompletionChunkModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'id': !(0, runtime_1.exists)(json, 'id') ? undefined : json['id'],
        'choices': !(0, runtime_1.exists)(json, 'choices') ? undefined : (json['choices'].map(ChoiceChunkModel_1.ChoiceChunkModelFromJSON)),
        'model': !(0, runtime_1.exists)(json, 'model') ? undefined : json['model'],
    };
}
function StreamChatCompletionChunkModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'id': value.id,
        'choices': value.choices === undefined ? undefined : (value.choices.map(ChoiceChunkModel_1.ChoiceChunkModelToJSON)),
        'model': value.model,
    };
}
