// @pinecone-database/pinecone/dist/pinecone-generated-ts-fetch/assistant_data/models/TextSnippetModel.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Pinecone Assistant Data Plane API
 * Pinecone Assistant Engine is a context engine to store and retrieve relevant knowledge from millions of documents at scale. This API supports interactions with assistants.
 *
 * The version of the OpenAPI document: 2025-10
 * Contact: support@pinecone.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.instanceOfTextSnippetModel = instanceOfTextSnippetModel;
exports.TextSnippetModelFromJSON = TextSnippetModelFromJSON;
exports.TextSnippetModelFromJSONTyped = TextSnippetModelFromJSONTyped;
exports.TextSnippetModelToJSON = TextSnippetModelToJSON;
const TypedReferenceModel_1 = require("./TypedReferenceModel");
/**
 * Check if a given object implements the TextSnippetModel interface.
 */
function instanceOfTextSnippetModel(value) {
    let isInstance = true;
    isInstance = isInstance && "type" in value;
    isInstance = isInstance && "content" in value;
    isInstance = isInstance && "score" in value;
    isInstance = isInstance && "reference" in value;
    return isInstance;
}
function TextSnippetModelFromJSON(json) {
    return TextSnippetModelFromJSONTyped(json, false);
}
function TextSnippetModelFromJSONTyped(json, ignoreDiscriminator) {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        'type': json['type'],
        'content': json['content'],
        'score': json['score'],
        'reference': (0, TypedReferenceModel_1.TypedReferenceModelFromJSON)(json['reference']),
    };
}
function TextSnippetModelToJSON(value) {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        'type': value.type,
        'content': value.content,
        'score': value.score,
        'reference': (0, TypedReferenceModel_1.TypedReferenceModelToJSON)(value.reference),
    };
}
